
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../algorithms/">
      
      
        <link rel="next" href="../data_processing/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.34">
    
    
      
        <title>Grape - ATHENA Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.35f28582.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#grape.grape" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="ATHENA Documentation" class="md-header__button md-logo" aria-label="ATHENA Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            ATHENA Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Grape
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="ATHENA Documentation" class="md-nav__button md-logo" aria-label="ATHENA Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    ATHENA Documentation
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Manual
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Manual
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manual/installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manual/parameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Parameters
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manual/input_files/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Input Files
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../manual/output_files/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Output Files
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methodology/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Methodology
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Reference
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_1" >
        
          
          <label class="md-nav__link" for="__nav_5_1" id="__nav_5_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    GENN
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_1">
            <span class="md-nav__icon md-icon"></span>
            GENN
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../alg_setup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Alg setup
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../functions/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Functions
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../parallel/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Parallel
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_2" checked>
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Grape
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            Grape
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../algorithms/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Algorithms
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Grape
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Grape
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#grape.grape" class="md-nav__link">
    <span class="md-ellipsis">
      grape
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.Grammar" class="md-nav__link">
    <span class="md-ellipsis">
      Grammar
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Grammar">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#grape.grape.Grammar.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#grape.grape.Grammar.check_recursiveness" class="md-nav__link">
    <span class="md-ellipsis">
      check_recursiveness
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.Individual" class="md-nav__link">
    <span class="md-ellipsis">
      Individual
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Individual">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#grape.grape.Individual.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.LeapIndividual" class="md-nav__link">
    <span class="md-ellipsis">
      LeapIndividual
    </span>
  </a>
  
    <nav class="md-nav" aria-label="LeapIndividual">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#grape.grape.LeapIndividual.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.MCGEIndividual" class="md-nav__link">
    <span class="md-ellipsis">
      MCGEIndividual
    </span>
  </a>
  
    <nav class="md-nav" aria-label="MCGEIndividual">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#grape.grape.MCGEIndividual.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.crossover_onepoint" class="md-nav__link">
    <span class="md-ellipsis">
      crossover_onepoint
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.leap_crossover_onepoint" class="md-nav__link">
    <span class="md-ellipsis">
      leap_crossover_onepoint
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.leap_mutation_int_flip_per_codon" class="md-nav__link">
    <span class="md-ellipsis">
      leap_mutation_int_flip_per_codon
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.leap_random_initialisation" class="md-nav__link">
    <span class="md-ellipsis">
      leap_random_initialisation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.leap_sensible_initialisation" class="md-nav__link">
    <span class="md-ellipsis">
      leap_sensible_initialisation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_eager" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_eager
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_eager_leap" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_eager_leap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_eager_mcge" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_eager_mcge
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_lazy" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_lazy
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_lazy_leap" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_lazy_leap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_lazy_mcge" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_lazy_mcge
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mcge_crossover_onepoint" class="md-nav__link">
    <span class="md-ellipsis">
      mcge_crossover_onepoint
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mcge_mutation_int_flip_per_codon" class="md-nav__link">
    <span class="md-ellipsis">
      mcge_mutation_int_flip_per_codon
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mcge_random_initializaion" class="md-nav__link">
    <span class="md-ellipsis">
      mcge_random_initializaion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mcge_sensible_initialisation" class="md-nav__link">
    <span class="md-ellipsis">
      mcge_sensible_initialisation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mutation_int_flip_per_codon" class="md-nav__link">
    <span class="md-ellipsis">
      mutation_int_flip_per_codon
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.random_initialisation" class="md-nav__link">
    <span class="md-ellipsis">
      random_initialisation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.reMap" class="md-nav__link">
    <span class="md-ellipsis">
      reMap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.reMap_leap" class="md-nav__link">
    <span class="md-ellipsis">
      reMap_leap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.reMap_mcge" class="md-nav__link">
    <span class="md-ellipsis">
      reMap_mcge
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.replace_nth" class="md-nav__link">
    <span class="md-ellipsis">
      replace_nth
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.selAthenaRoulette" class="md-nav__link">
    <span class="md-ellipsis">
      selAthenaRoulette
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.selAutoEpsilonLexicase" class="md-nav__link">
    <span class="md-ellipsis">
      selAutoEpsilonLexicase
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.selBalAccLexicase" class="md-nav__link">
    <span class="md-ellipsis">
      selBalAccLexicase
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.selTournRoulette" class="md-nav__link">
    <span class="md-ellipsis">
      selTournRoulette
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.selTournamentWithoutInvalids" class="md-nav__link">
    <span class="md-ellipsis">
      selTournamentWithoutInvalids
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.sensible_initialisation" class="md-nav__link">
    <span class="md-ellipsis">
      sensible_initialisation
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5_3" >
        
          
          <label class="md-nav__link" for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Utilites
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            Utilites
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../data_processing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data processing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../parameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Parameters
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#grape.grape" class="md-nav__link">
    <span class="md-ellipsis">
      grape
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.Grammar" class="md-nav__link">
    <span class="md-ellipsis">
      Grammar
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Grammar">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#grape.grape.Grammar.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#grape.grape.Grammar.check_recursiveness" class="md-nav__link">
    <span class="md-ellipsis">
      check_recursiveness
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.Individual" class="md-nav__link">
    <span class="md-ellipsis">
      Individual
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Individual">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#grape.grape.Individual.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.LeapIndividual" class="md-nav__link">
    <span class="md-ellipsis">
      LeapIndividual
    </span>
  </a>
  
    <nav class="md-nav" aria-label="LeapIndividual">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#grape.grape.LeapIndividual.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.MCGEIndividual" class="md-nav__link">
    <span class="md-ellipsis">
      MCGEIndividual
    </span>
  </a>
  
    <nav class="md-nav" aria-label="MCGEIndividual">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#grape.grape.MCGEIndividual.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.crossover_onepoint" class="md-nav__link">
    <span class="md-ellipsis">
      crossover_onepoint
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.leap_crossover_onepoint" class="md-nav__link">
    <span class="md-ellipsis">
      leap_crossover_onepoint
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.leap_mutation_int_flip_per_codon" class="md-nav__link">
    <span class="md-ellipsis">
      leap_mutation_int_flip_per_codon
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.leap_random_initialisation" class="md-nav__link">
    <span class="md-ellipsis">
      leap_random_initialisation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.leap_sensible_initialisation" class="md-nav__link">
    <span class="md-ellipsis">
      leap_sensible_initialisation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_eager" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_eager
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_eager_leap" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_eager_leap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_eager_mcge" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_eager_mcge
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_lazy" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_lazy
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_lazy_leap" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_lazy_leap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mapper_lazy_mcge" class="md-nav__link">
    <span class="md-ellipsis">
      mapper_lazy_mcge
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mcge_crossover_onepoint" class="md-nav__link">
    <span class="md-ellipsis">
      mcge_crossover_onepoint
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mcge_mutation_int_flip_per_codon" class="md-nav__link">
    <span class="md-ellipsis">
      mcge_mutation_int_flip_per_codon
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mcge_random_initializaion" class="md-nav__link">
    <span class="md-ellipsis">
      mcge_random_initializaion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mcge_sensible_initialisation" class="md-nav__link">
    <span class="md-ellipsis">
      mcge_sensible_initialisation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.mutation_int_flip_per_codon" class="md-nav__link">
    <span class="md-ellipsis">
      mutation_int_flip_per_codon
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.random_initialisation" class="md-nav__link">
    <span class="md-ellipsis">
      random_initialisation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.reMap" class="md-nav__link">
    <span class="md-ellipsis">
      reMap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.reMap_leap" class="md-nav__link">
    <span class="md-ellipsis">
      reMap_leap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.reMap_mcge" class="md-nav__link">
    <span class="md-ellipsis">
      reMap_mcge
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.replace_nth" class="md-nav__link">
    <span class="md-ellipsis">
      replace_nth
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.selAthenaRoulette" class="md-nav__link">
    <span class="md-ellipsis">
      selAthenaRoulette
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.selAutoEpsilonLexicase" class="md-nav__link">
    <span class="md-ellipsis">
      selAutoEpsilonLexicase
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.selBalAccLexicase" class="md-nav__link">
    <span class="md-ellipsis">
      selBalAccLexicase
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.selTournRoulette" class="md-nav__link">
    <span class="md-ellipsis">
      selTournRoulette
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.selTournamentWithoutInvalids" class="md-nav__link">
    <span class="md-ellipsis">
      selTournamentWithoutInvalids
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grape.grape.sensible_initialisation" class="md-nav__link">
    <span class="md-ellipsis">
      sensible_initialisation
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>Grape</h1>

<p>Provides classes and functiosn for the <code>ATHENA</code> project 
providing for alternative genome types with corresponding
mapping and initialization approaches. Provides
alternative selection types for evolutionary algorithm.</p>


<div class="doc doc-object doc-module">



<a id="grape.grape"></a>
    <div class="doc doc-contents first">

        <p>Modified code from GRAPE: Grammatical Algorithms in Python for Evolution for use with ATHENA and includes multi-chromosome and LEAP genome implementations</p>



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="grape.grape.Grammar" class="doc doc-heading">
            <code>Grammar</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code>object</code></p>


        <p>BNF Grammar 
Attributes:
    non_terminals: list with each non-terminal (NT);
    start_rule: first non-terminal;
    nt_rule_size: number of nt rules used for frame size for set of codons in leap mapping 
        and for the number of chromosomes in MCGE
    production_rules: list with each production rule (PR), which contains in each position:
        - the PR itself as a string
        - 'non-terminal' or 'terminal'
        - the arity (number of NTs in the PR)
        - production choice label
        - True, if it is recursive, and False, otherwise
        - the minimum depth to terminate the mapping of all NTs of this PR
    n_rules: df</p>

              <details class="quote">
                <summary>Source code in <code>grape/grape.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Grammar</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    BNF Grammar </span>
<span class="sd">    Attributes:</span>
<span class="sd">        non_terminals: list with each non-terminal (NT);</span>
<span class="sd">        start_rule: first non-terminal;</span>
<span class="sd">        nt_rule_size: number of nt rules used for frame size for set of codons in leap mapping </span>
<span class="sd">            and for the number of chromosomes in MCGE</span>
<span class="sd">        production_rules: list with each production rule (PR), which contains in each position:</span>
<span class="sd">            - the PR itself as a string</span>
<span class="sd">            - &#39;non-terminal&#39; or &#39;terminal&#39;</span>
<span class="sd">            - the arity (number of NTs in the PR)</span>
<span class="sd">            - production choice label</span>
<span class="sd">            - True, if it is recursive, and False, otherwise</span>
<span class="sd">            - the minimum depth to terminate the mapping of all NTs of this PR</span>
<span class="sd">        n_rules: df</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span><span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span><span class="s1">&#39;eager&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot; Initializes Grammar instance by reading BNF grammar string </span>
<span class="sd">        Args:</span>
<span class="sd">            bnf_grammar: text describing BNF grammar</span>
<span class="sd">            codon_consumption: &#39;eager&#39; or &#39;lazy&#39; controlling consumption during mapping</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bnf_grammar</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">)</span>

        <span class="c1">#self.non_terminals = [&#39;&lt;&#39; + term + &#39;&gt;&#39; for term in re.findall(r&quot;\&lt;(\w+)\&gt;\s*::=&quot;,bnf_grammar)]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;\s*::=&quot;</span><span class="p">,</span><span class="n">bnf_grammar</span><span class="p">)]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">start_rule</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">)):</span>
            <span class="n">bnf_grammar</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot; ::=&quot;</span><span class="p">,</span> <span class="s2">&quot;  ::=&quot;</span><span class="p">)</span>
        <span class="n">rules</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;::=&quot;</span><span class="p">)</span>
        <span class="k">del</span> <span class="n">rules</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">rules</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">rules</span><span class="p">]</span>
        <span class="n">rules</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">rules</span><span class="p">]</span>

        <span class="c1">#list of lists (set of production rules for each non-terminal)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">rules</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">)):</span>
            <span class="c1">#Getting rid of all leading and trailing whitespaces</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">])):</span>
                <span class="c1">#Include in the list the PR itself, NT or T, arity and the production choice label</span>
                <span class="c1">#if re.findall(r&quot;\&lt;(\w+)\&gt;&quot;,self.production_rules[i][j]):</span>
                <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]):</span>                    
                    <span class="c1">#arity = len(re.findall(r&quot;\&lt;(\w+)\&gt;&quot;,self.production_rules[i][j]))</span>
                    <span class="n">arity</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]))</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="p">,</span> <span class="s2">&quot;non-terminal&quot;</span><span class="p">,</span> <span class="n">arity</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="p">,</span> <span class="s2">&quot;terminal&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="c1">#arity 0</span>
        <span class="c1">#number of production rules for each non-terminal</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_rules</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">list_</span><span class="p">)</span> <span class="k">for</span> <span class="n">list_</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">])):</span>
                <span class="n">NTs_to_check_recursiveness</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">NTs_to_check_recursiveness</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">item_</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">item_</span> <span class="ow">in</span> <span class="n">NTs_to_check_recursiveness</span><span class="p">]</span>
                <span class="n">unique_NTs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">NTs_to_check_recursiveness</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> 
                <span class="n">recursive</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">for</span> <span class="n">NT_to_check</span> <span class="ow">in</span> <span class="n">unique_NTs</span><span class="p">:</span>
                    <span class="n">stack</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>  
                    <span class="k">if</span> <span class="n">NT_to_check</span> <span class="ow">in</span> <span class="n">stack</span><span class="p">:</span>
                        <span class="n">recursive</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="k">break</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">stack</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">NT_to_check</span><span class="p">)</span>
                        <span class="n">recursive</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_recursiveness</span><span class="p">(</span><span class="n">NT_to_check</span><span class="p">,</span> <span class="n">stack</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">recursive</span><span class="p">:</span>
                            <span class="k">break</span>
                        <span class="n">stack</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">recursive</span><span class="p">)</span>

        <span class="c1">#minimum depth from each non-terminal to terminate the mapping of all symbols</span>
        <span class="n">NT_depth_to_terminate</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">)</span>
        <span class="c1">#minimum depth from each production rule to terminate the mapping of all symbols</span>
        <span class="n">part_PR_depth_to_terminate</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span> <span class="c1">#min depth for each non-terminal or terminal to terminate</span>
        <span class="n">isolated_non_terminal</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span> <span class="c1">#None, if the respective position has a terminal</span>
        <span class="c1">#Separating the non-terminals within the same production rule</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">)):</span>
            <span class="n">part_PR_depth_to_terminate</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="nb">list</span><span class="p">()</span> <span class="p">)</span>
            <span class="n">isolated_non_terminal</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="nb">list</span><span class="p">()</span> <span class="p">)</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">])):</span>
                <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="nb">list</span><span class="p">()</span> <span class="p">)</span>
                <span class="n">isolated_non_terminal</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="nb">list</span><span class="p">()</span> <span class="p">)</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;terminal&#39;</span><span class="p">:</span>
                    <span class="n">isolated_non_terminal</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                    <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
                        <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">][</span><span class="mi">2</span><span class="p">]):</span> <span class="c1">#arity</span>
                        <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="nb">list</span><span class="p">()</span> <span class="p">)</span>
                        <span class="c1">#term = re.findall(r&quot;\&lt;(\w+)\&gt;&quot;,self.production_rules[i][j][0])[k]</span>
                        <span class="n">term</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">])[</span><span class="n">k</span><span class="p">]</span>
                        <span class="n">isolated_non_terminal</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">)</span>
        <span class="n">continue_</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">while</span> <span class="n">continue_</span><span class="p">:</span>
            <span class="c1">#after filling up NT_depth_to_terminate, we need to run the loop one more time to</span>
            <span class="c1">#fill up part_PR_depth_to_terminate, so we check in the beginning</span>
            <span class="k">if</span> <span class="kc">None</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">NT_depth_to_terminate</span><span class="p">:</span>
                <span class="n">continue_</span> <span class="o">=</span> <span class="kc">False</span> 
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">)):</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">)):</span>
                    <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">j</span><span class="p">])):</span>
                        <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">isolated_non_terminal</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">])):</span>
                            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">isolated_non_terminal</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">][</span><span class="n">l</span><span class="p">]:</span>
                                <span class="k">if</span> <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
                                    <span class="k">if</span> <span class="ow">not</span> <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">][</span><span class="n">l</span><span class="p">]:</span>
                                        <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">][</span><span class="n">l</span><span class="p">]</span> <span class="o">=</span> <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
                                        <span class="k">if</span> <span class="p">[]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">]:</span>
                                            <span class="k">if</span> <span class="ow">not</span> <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
                                                <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">][</span><span class="n">l</span><span class="p">]</span>
        <span class="n">PR_depth_to_terminate</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">part_PR_depth_to_terminate</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">])):</span>
                <span class="c1">#the min depth to terminate a PR is the max depth within the items of that PR</span>
                <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">])</span> <span class="o">==</span> <span class="nb">int</span><span class="p">:</span>
                    <span class="n">depth_</span> <span class="o">=</span> <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span>
                    <span class="n">PR_depth_to_terminate</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">depth_</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">depth_</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">depth_</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">])</span>
                    <span class="n">PR_depth_to_terminate</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">depth_</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">depth_</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">rule_conversion</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">idx</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nt_rule_size</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">)):</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">rule_conversion</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">nt_rule_size</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">idx</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">rule_conversion</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span> <span class="c1">#codon consumption is lazy</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">rule_conversion</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">))]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">nt_rule_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">check_recursiveness</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">NT</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">str_</span><span class="p">,</span> <span class="n">stack</span><span class="p">:</span> <span class="nb">list</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot; Determines whether a non-terminal in the grammar is recursive </span>
<span class="sd">        Args:</span>
<span class="sd">            NT: non-terminal checked</span>
<span class="sd">            stack: tracks non-terminals traversed</span>

<span class="sd">        Returns:</span>
<span class="sd">            True is recursive, False if not</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">idx_NT</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">NT</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">])):</span>
            <span class="n">NTs_to_check_recursiveness</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">][</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">NTs_to_check_recursiveness</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">item_</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">item_</span> <span class="ow">in</span> <span class="n">NTs_to_check_recursiveness</span><span class="p">]</span>
            <span class="n">unique_NTs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">NTs_to_check_recursiveness</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> 
            <span class="n">recursive</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="c1">#      while unique_NTs.size and not recursive:</span>
            <span class="k">for</span> <span class="n">NT_to_check</span> <span class="ow">in</span> <span class="n">unique_NTs</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">NT_to_check</span> <span class="ow">in</span> <span class="n">stack</span><span class="p">:</span>
                    <span class="n">recursive</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">return</span> <span class="n">recursive</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">stack</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">NT_to_check</span><span class="p">)</span> <span class="c1">#Include the current NT to check it recursively</span>
                    <span class="n">recursive</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_recursiveness</span><span class="p">(</span><span class="n">NT_to_check</span><span class="p">,</span> <span class="n">stack</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">recursive</span><span class="p">:</span>
                        <span class="k">return</span> <span class="n">recursive</span>
                    <span class="n">stack</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span> <span class="c1">#If the inclusion didn&#39;t show recursiveness, remove it before continuing</span>
        <span class="k">return</span> <span class="n">recursive</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="grape.grape.Grammar.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="o">=</span><span class="s1">&#39;eager&#39;</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Initializes Grammar instance by reading BNF grammar string 
Args:
    bnf_grammar: text describing BNF grammar
    codon_consumption: 'eager' or 'lazy' controlling consumption during mapping</p>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span><span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span> <span class="o">=</span><span class="s1">&#39;eager&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; Initializes Grammar instance by reading BNF grammar string </span>
<span class="sd">    Args:</span>
<span class="sd">        bnf_grammar: text describing BNF grammar</span>
<span class="sd">        codon_consumption: &#39;eager&#39; or &#39;lazy&#39; controlling consumption during mapping</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bnf_grammar</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">)</span>

    <span class="c1">#self.non_terminals = [&#39;&lt;&#39; + term + &#39;&gt;&#39; for term in re.findall(r&quot;\&lt;(\w+)\&gt;\s*::=&quot;,bnf_grammar)]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;\s*::=&quot;</span><span class="p">,</span><span class="n">bnf_grammar</span><span class="p">)]</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">start_rule</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">)):</span>
        <span class="n">bnf_grammar</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot; ::=&quot;</span><span class="p">,</span> <span class="s2">&quot;  ::=&quot;</span><span class="p">)</span>
    <span class="n">rules</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;::=&quot;</span><span class="p">)</span>
    <span class="k">del</span> <span class="n">rules</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">rules</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">rules</span><span class="p">]</span>
    <span class="n">rules</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">rules</span><span class="p">]</span>

    <span class="c1">#list of lists (set of production rules for each non-terminal)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">rules</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">)):</span>
        <span class="c1">#Getting rid of all leading and trailing whitespaces</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">])):</span>
            <span class="c1">#Include in the list the PR itself, NT or T, arity and the production choice label</span>
            <span class="c1">#if re.findall(r&quot;\&lt;(\w+)\&gt;&quot;,self.production_rules[i][j]):</span>
            <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]):</span>                    
                <span class="c1">#arity = len(re.findall(r&quot;\&lt;(\w+)\&gt;&quot;,self.production_rules[i][j]))</span>
                <span class="n">arity</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]))</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="p">,</span> <span class="s2">&quot;non-terminal&quot;</span><span class="p">,</span> <span class="n">arity</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="p">,</span> <span class="s2">&quot;terminal&quot;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="c1">#arity 0</span>
    <span class="c1">#number of production rules for each non-terminal</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">n_rules</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">list_</span><span class="p">)</span> <span class="k">for</span> <span class="n">list_</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">])):</span>
            <span class="n">NTs_to_check_recursiveness</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">NTs_to_check_recursiveness</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">item_</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">item_</span> <span class="ow">in</span> <span class="n">NTs_to_check_recursiveness</span><span class="p">]</span>
            <span class="n">unique_NTs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">NTs_to_check_recursiveness</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> 
            <span class="n">recursive</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">for</span> <span class="n">NT_to_check</span> <span class="ow">in</span> <span class="n">unique_NTs</span><span class="p">:</span>
                <span class="n">stack</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>  
                <span class="k">if</span> <span class="n">NT_to_check</span> <span class="ow">in</span> <span class="n">stack</span><span class="p">:</span>
                    <span class="n">recursive</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">break</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">stack</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">NT_to_check</span><span class="p">)</span>
                    <span class="n">recursive</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_recursiveness</span><span class="p">(</span><span class="n">NT_to_check</span><span class="p">,</span> <span class="n">stack</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">recursive</span><span class="p">:</span>
                        <span class="k">break</span>
                    <span class="n">stack</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">recursive</span><span class="p">)</span>

    <span class="c1">#minimum depth from each non-terminal to terminate the mapping of all symbols</span>
    <span class="n">NT_depth_to_terminate</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">)</span>
    <span class="c1">#minimum depth from each production rule to terminate the mapping of all symbols</span>
    <span class="n">part_PR_depth_to_terminate</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span> <span class="c1">#min depth for each non-terminal or terminal to terminate</span>
    <span class="n">isolated_non_terminal</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span> <span class="c1">#None, if the respective position has a terminal</span>
    <span class="c1">#Separating the non-terminals within the same production rule</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">)):</span>
        <span class="n">part_PR_depth_to_terminate</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="nb">list</span><span class="p">()</span> <span class="p">)</span>
        <span class="n">isolated_non_terminal</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="nb">list</span><span class="p">()</span> <span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">])):</span>
            <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="nb">list</span><span class="p">()</span> <span class="p">)</span>
            <span class="n">isolated_non_terminal</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="nb">list</span><span class="p">()</span> <span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;terminal&#39;</span><span class="p">:</span>
                <span class="n">isolated_non_terminal</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
                <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
                    <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">][</span><span class="mi">2</span><span class="p">]):</span> <span class="c1">#arity</span>
                    <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span> <span class="nb">list</span><span class="p">()</span> <span class="p">)</span>
                    <span class="c1">#term = re.findall(r&quot;\&lt;(\w+)\&gt;&quot;,self.production_rules[i][j][0])[k]</span>
                    <span class="n">term</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">])[</span><span class="n">k</span><span class="p">]</span>
                    <span class="n">isolated_non_terminal</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span><span class="p">)</span>
    <span class="n">continue_</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">while</span> <span class="n">continue_</span><span class="p">:</span>
        <span class="c1">#after filling up NT_depth_to_terminate, we need to run the loop one more time to</span>
        <span class="c1">#fill up part_PR_depth_to_terminate, so we check in the beginning</span>
        <span class="k">if</span> <span class="kc">None</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">NT_depth_to_terminate</span><span class="p">:</span>
            <span class="n">continue_</span> <span class="o">=</span> <span class="kc">False</span> 
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">)):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">)):</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">j</span><span class="p">])):</span>
                    <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">isolated_non_terminal</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">])):</span>
                        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">isolated_non_terminal</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">][</span><span class="n">l</span><span class="p">]:</span>
                            <span class="k">if</span> <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
                                <span class="k">if</span> <span class="ow">not</span> <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">][</span><span class="n">l</span><span class="p">]:</span>
                                    <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">][</span><span class="n">l</span><span class="p">]</span> <span class="o">=</span> <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span>
                                    <span class="k">if</span> <span class="p">[]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">]:</span>
                                        <span class="k">if</span> <span class="ow">not</span> <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">]:</span>
                                            <span class="n">NT_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">j</span><span class="p">][</span><span class="n">k</span><span class="p">][</span><span class="n">l</span><span class="p">]</span>
    <span class="n">PR_depth_to_terminate</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">part_PR_depth_to_terminate</span><span class="p">)):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">])):</span>
            <span class="c1">#the min depth to terminate a PR is the max depth within the items of that PR</span>
            <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">])</span> <span class="o">==</span> <span class="nb">int</span><span class="p">:</span>
                <span class="n">depth_</span> <span class="o">=</span> <span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span>
                <span class="n">PR_depth_to_terminate</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">depth_</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">depth_</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">depth_</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">part_PR_depth_to_terminate</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">])</span>
                <span class="n">PR_depth_to_terminate</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">depth_</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">j</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">depth_</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">rule_conversion</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nt_rule_size</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">)):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">rule_conversion</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">nt_rule_size</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">idx</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">rule_conversion</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span> <span class="c1">#codon consumption is lazy</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rule_conversion</span><span class="o">=</span><span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="p">))]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nt_rule_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="grape.grape.Grammar.check_recursiveness" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">check_recursiveness</span><span class="p">(</span><span class="n">NT</span><span class="p">,</span> <span class="n">stack</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Determines whether a non-terminal in the grammar is recursive 
Args:
    NT: non-terminal checked
    stack: tracks non-terminals traversed</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True is recursive, False if not</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">check_recursiveness</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">NT</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">str_</span><span class="p">,</span> <span class="n">stack</span><span class="p">:</span> <span class="nb">list</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; Determines whether a non-terminal in the grammar is recursive </span>
<span class="sd">    Args:</span>
<span class="sd">        NT: non-terminal checked</span>
<span class="sd">        stack: tracks non-terminals traversed</span>

<span class="sd">    Returns:</span>
<span class="sd">        True is recursive, False if not</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">idx_NT</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">NT</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">])):</span>
        <span class="n">NTs_to_check_recursiveness</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">][</span><span class="n">j</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">NTs_to_check_recursiveness</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">item_</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">item_</span> <span class="ow">in</span> <span class="n">NTs_to_check_recursiveness</span><span class="p">]</span>
        <span class="n">unique_NTs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">NTs_to_check_recursiveness</span><span class="p">,</span> <span class="n">return_counts</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> 
        <span class="n">recursive</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1">#      while unique_NTs.size and not recursive:</span>
        <span class="k">for</span> <span class="n">NT_to_check</span> <span class="ow">in</span> <span class="n">unique_NTs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">NT_to_check</span> <span class="ow">in</span> <span class="n">stack</span><span class="p">:</span>
                <span class="n">recursive</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">return</span> <span class="n">recursive</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">stack</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">NT_to_check</span><span class="p">)</span> <span class="c1">#Include the current NT to check it recursively</span>
                <span class="n">recursive</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">check_recursiveness</span><span class="p">(</span><span class="n">NT_to_check</span><span class="p">,</span> <span class="n">stack</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">recursive</span><span class="p">:</span>
                    <span class="k">return</span> <span class="n">recursive</span>
                <span class="n">stack</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span> <span class="c1">#If the inclusion didn&#39;t show recursiveness, remove it before continuing</span>
    <span class="k">return</span> <span class="n">recursive</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="grape.grape.Individual" class="doc doc-heading">
            <code>Individual</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code>object</code></p>


        <p>A GE individual.</p>

              <details class="quote">
                <summary>Source code in <code>grape/grape.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">Individual</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A GE individual.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>
        <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_lazy</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_eager</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
<span class="c1">#             self.structure = mapper_eager_opt(genome, grammar, max_depth)</span>


        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="grape.grape.Individual.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span></pre></div></td><td class="code"><div><pre><span></span><code>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>
        <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_lazy</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_eager</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
<span class="c1">#             self.structure = mapper_eager_opt(genome, grammar, max_depth)</span>


        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="grape.grape.LeapIndividual" class="doc doc-heading">
            <code>LeapIndividual</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code>object</code></p>


        <p>A GE LEAP mapping individual.
genome is divided into frames - one codon per non-terminal in a frame
stay in frame until need another NT that has already been consumed</p>

              <details class="quote">
                <summary>Source code in <code>grape/grape.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">LeapIndividual</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A GE LEAP mapping individual.</span>
<span class="sd">    genome is divided into frames - one codon per non-terminal in a frame</span>
<span class="sd">    stay in frame until need another NT that has already been consumed</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>


        <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_lazy_leap</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_eager_leap</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="grape.grape.LeapIndividual.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>


    <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_lazy_leap</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_eager_leap</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h2 id="grape.grape.MCGEIndividual" class="doc doc-heading">
            <code>MCGEIndividual</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code>object</code></p>


        <p>A  multichromosome GE mapping individual.
genome is divided into chromosomes
each chromosome contains codons for a non-terminal in the grammar</p>

              <details class="quote">
                <summary>Source code in <code>grape/grape.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">MCGEIndividual</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A  multichromosome GE mapping individual.</span>
<span class="sd">    genome is divided into chromosomes</span>
<span class="sd">    each chromosome contains codons for a non-terminal in the grammar</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>
        <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">consumed_codons</span> <span class="o">=</span> <span class="n">mapper_lazy_mcge</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
            <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">consumed_codons</span> <span class="o">=</span> <span class="n">mapper_eager_mcge</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="grape.grape.MCGEIndividual.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>
    <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">consumed_codons</span> <span class="o">=</span> <span class="n">mapper_lazy_mcge</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
        <span class="bp">self</span><span class="o">.</span><span class="n">structure</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">consumed_codons</span> <span class="o">=</span> <span class="n">mapper_eager_mcge</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h2 id="grape.grape.crossover_onepoint" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">crossover_onepoint</span><span class="p">(</span><span class="n">parent0</span><span class="p">,</span> <span class="n">parent1</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">genome_representation</span><span class="o">=</span><span class="s1">&#39;list&#39;</span><span class="p">,</span> <span class="n">max_genome_length</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>One point crossover for standard linear GE genome individuals</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>parent0</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>individual to cross</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>parent1</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>second individual to cross</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genome_representation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>either list or numpy</p>
              </div>
            </td>
            <td>
                  <code>&#39;list&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum allowed length of genome when not None</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>new_ind0</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new individual resulting from crossover</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>new_ind1</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>second new individual resulting from crossover</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1578</span>
<span class="normal">1579</span>
<span class="normal">1580</span>
<span class="normal">1581</span>
<span class="normal">1582</span>
<span class="normal">1583</span>
<span class="normal">1584</span>
<span class="normal">1585</span>
<span class="normal">1586</span>
<span class="normal">1587</span>
<span class="normal">1588</span>
<span class="normal">1589</span>
<span class="normal">1590</span>
<span class="normal">1591</span>
<span class="normal">1592</span>
<span class="normal">1593</span>
<span class="normal">1594</span>
<span class="normal">1595</span>
<span class="normal">1596</span>
<span class="normal">1597</span>
<span class="normal">1598</span>
<span class="normal">1599</span>
<span class="normal">1600</span>
<span class="normal">1601</span>
<span class="normal">1602</span>
<span class="normal">1603</span>
<span class="normal">1604</span>
<span class="normal">1605</span>
<span class="normal">1606</span>
<span class="normal">1607</span>
<span class="normal">1608</span>
<span class="normal">1609</span>
<span class="normal">1610</span>
<span class="normal">1611</span>
<span class="normal">1612</span>
<span class="normal">1613</span>
<span class="normal">1614</span>
<span class="normal">1615</span>
<span class="normal">1616</span>
<span class="normal">1617</span>
<span class="normal">1618</span>
<span class="normal">1619</span>
<span class="normal">1620</span>
<span class="normal">1621</span>
<span class="normal">1622</span>
<span class="normal">1623</span>
<span class="normal">1624</span>
<span class="normal">1625</span>
<span class="normal">1626</span>
<span class="normal">1627</span>
<span class="normal">1628</span>
<span class="normal">1629</span>
<span class="normal">1630</span>
<span class="normal">1631</span>
<span class="normal">1632</span>
<span class="normal">1633</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">crossover_onepoint</span><span class="p">(</span><span class="n">parent0</span><span class="p">:</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> <span class="n">parent1</span><span class="p">:</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> 
                       <span class="n">bnf_grammar</span><span class="p">:</span> <span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
                    <span class="n">genome_representation</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">&#39;list&#39;</span><span class="p">,</span> 
                    <span class="n">max_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span><span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;One point crossover for standard linear GE genome individuals</span>

<span class="sd">    Args:</span>
<span class="sd">        parent0: individual to cross</span>
<span class="sd">        parent1: second individual to cross</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        genome_representation: either list or numpy</span>
<span class="sd">        max_genome_length: maximum allowed length of genome when not None</span>

<span class="sd">    Returns:</span>
<span class="sd">        new_ind0: new individual resulting from crossover</span>
<span class="sd">        new_ind1: second new individual resulting from crossover</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="c1"># restrict crossover to effective genome when individual is valid</span>
    <span class="k">if</span> <span class="n">parent0</span><span class="o">.</span><span class="n">invalid</span><span class="p">:</span> <span class="c1">#used_codons = 0</span>
        <span class="n">possible_crossover_codons0</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">possible_crossover_codons0</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">),</span> <span class="n">parent0</span><span class="o">.</span><span class="n">used_codons</span><span class="p">)</span> <span class="c1">#in case of wrapping, used_codons can be greater than genome&#39;s length</span>
    <span class="k">if</span> <span class="n">parent1</span><span class="o">.</span><span class="n">invalid</span><span class="p">:</span>
        <span class="n">possible_crossover_codons1</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">possible_crossover_codons1</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">),</span> <span class="n">parent1</span><span class="o">.</span><span class="n">used_codons</span><span class="p">)</span>

    <span class="n">point0</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">possible_crossover_codons0</span><span class="p">)</span>
    <span class="n">point1</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">possible_crossover_codons1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
        <span class="c1">#Operate crossover</span>
        <span class="n">new_genome0</span> <span class="o">=</span> <span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point0</span><span class="p">]</span> <span class="o">+</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">point1</span><span class="p">:]</span>
        <span class="n">new_genome1</span> <span class="o">=</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point1</span><span class="p">]</span> <span class="o">+</span> <span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">point0</span><span class="p">:]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Only &#39;list&#39; representation is implemented&quot;</span><span class="p">)</span>

    <span class="n">new_ind0</span> <span class="o">=</span> <span class="n">reMap</span><span class="p">(</span><span class="n">parent0</span><span class="p">,</span> <span class="n">new_genome0</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>
    <span class="n">new_ind1</span> <span class="o">=</span> <span class="n">reMap</span><span class="p">(</span><span class="n">parent1</span><span class="p">,</span> <span class="n">new_genome1</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">new_ind0</span><span class="o">.</span><span class="n">depth</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
        <span class="n">new_ind0</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">if</span> <span class="n">new_ind1</span><span class="o">.</span><span class="n">depth</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
        <span class="n">new_ind1</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">max_genome_length</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_ind0</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_genome_length</span><span class="p">:</span>
            <span class="n">new_ind0</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_ind1</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_genome_length</span><span class="p">:</span>
            <span class="n">new_ind1</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">del</span> <span class="n">new_ind0</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">new_ind1</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span>
    <span class="k">return</span> <span class="n">new_ind0</span><span class="p">,</span> <span class="n">new_ind1</span>   
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.leap_crossover_onepoint" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">leap_crossover_onepoint</span><span class="p">(</span><span class="n">parent0</span><span class="p">,</span> <span class="n">parent1</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">genome_representation</span><span class="o">=</span><span class="s1">&#39;list&#39;</span><span class="p">,</span> <span class="n">max_genome_length</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>One point crossover for standard LEAP GE genome individuals. Crossovers occur
    between frames</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>parent0</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>individual to cross</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>parent1</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>second individual to cross</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genome_representation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>either list or numpy</p>
              </div>
            </td>
            <td>
                  <code>&#39;list&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum allowed length of genome when not None</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>new_ind0</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new individual resulting from crossover</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>new_ind1</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>second new individual resulting from crossover</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1637</span>
<span class="normal">1638</span>
<span class="normal">1639</span>
<span class="normal">1640</span>
<span class="normal">1641</span>
<span class="normal">1642</span>
<span class="normal">1643</span>
<span class="normal">1644</span>
<span class="normal">1645</span>
<span class="normal">1646</span>
<span class="normal">1647</span>
<span class="normal">1648</span>
<span class="normal">1649</span>
<span class="normal">1650</span>
<span class="normal">1651</span>
<span class="normal">1652</span>
<span class="normal">1653</span>
<span class="normal">1654</span>
<span class="normal">1655</span>
<span class="normal">1656</span>
<span class="normal">1657</span>
<span class="normal">1658</span>
<span class="normal">1659</span>
<span class="normal">1660</span>
<span class="normal">1661</span>
<span class="normal">1662</span>
<span class="normal">1663</span>
<span class="normal">1664</span>
<span class="normal">1665</span>
<span class="normal">1666</span>
<span class="normal">1667</span>
<span class="normal">1668</span>
<span class="normal">1669</span>
<span class="normal">1670</span>
<span class="normal">1671</span>
<span class="normal">1672</span>
<span class="normal">1673</span>
<span class="normal">1674</span>
<span class="normal">1675</span>
<span class="normal">1676</span>
<span class="normal">1677</span>
<span class="normal">1678</span>
<span class="normal">1679</span>
<span class="normal">1680</span>
<span class="normal">1681</span>
<span class="normal">1682</span>
<span class="normal">1683</span>
<span class="normal">1684</span>
<span class="normal">1685</span>
<span class="normal">1686</span>
<span class="normal">1687</span>
<span class="normal">1688</span>
<span class="normal">1689</span>
<span class="normal">1690</span>
<span class="normal">1691</span>
<span class="normal">1692</span>
<span class="normal">1693</span>
<span class="normal">1694</span>
<span class="normal">1695</span>
<span class="normal">1696</span>
<span class="normal">1697</span>
<span class="normal">1698</span>
<span class="normal">1699</span>
<span class="normal">1700</span>
<span class="normal">1701</span>
<span class="normal">1702</span>
<span class="normal">1703</span>
<span class="normal">1704</span>
<span class="normal">1705</span>
<span class="normal">1706</span>
<span class="normal">1707</span>
<span class="normal">1708</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">leap_crossover_onepoint</span><span class="p">(</span><span class="n">parent0</span><span class="p">:</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> <span class="n">parent1</span><span class="p">:</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> 
                       <span class="n">bnf_grammar</span><span class="p">:</span> <span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
                    <span class="n">genome_representation</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">&#39;list&#39;</span><span class="p">,</span> 
                    <span class="n">max_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span><span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;One point crossover for standard LEAP GE genome individuals. Crossovers occur</span>
<span class="sd">        between frames</span>

<span class="sd">    Args:</span>
<span class="sd">        parent0: individual to cross</span>
<span class="sd">        parent1: second individual to cross</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        genome_representation: either list or numpy</span>
<span class="sd">        max_genome_length: maximum allowed length of genome when not None</span>

<span class="sd">    Returns:</span>
<span class="sd">        new_ind0: new individual resulting from crossover</span>
<span class="sd">        new_ind1: second new individual resulting from crossover</span>
<span class="sd">    &quot;&quot;&quot;</span> 
    <span class="k">if</span> <span class="n">parent0</span><span class="o">.</span><span class="n">invalid</span><span class="p">:</span> <span class="c1">#used_codons = 0</span>
        <span class="n">possible_crossover_codons0</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">possible_crossover_codons0</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">),</span> <span class="n">parent0</span><span class="o">.</span><span class="n">used_codons</span><span class="p">)</span> <span class="c1">#in case of wrapping, used_codons can be greater than genome&#39;s length</span>
    <span class="k">if</span> <span class="n">parent1</span><span class="o">.</span><span class="n">invalid</span><span class="p">:</span>
        <span class="n">possible_crossover_codons1</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">possible_crossover_codons1</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">),</span> <span class="n">parent1</span><span class="o">.</span><span class="n">used_codons</span><span class="p">)</span>

    <span class="n">frame0</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">possible_crossover_codons0</span> <span class="o">//</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span> <span class="o">-</span><span class="mi">1</span> <span class="p">)</span>
    <span class="n">frame1</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">possible_crossover_codons1</span> <span class="o">//</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span> <span class="o">-</span><span class="mi">1</span> <span class="p">)</span>

    <span class="n">codon_cross</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span>

    <span class="n">point0</span> <span class="o">=</span> <span class="n">frame0</span><span class="o">*</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="o">+</span><span class="n">codon_cross</span>
    <span class="n">point1</span> <span class="o">=</span> <span class="n">frame1</span><span class="o">*</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="o">+</span><span class="n">codon_cross</span>

    <span class="k">if</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;numpy&#39;</span><span class="p">:</span>
        <span class="c1">#TODO This operations is not working in case of wrapping</span>
        <span class="n">len0</span> <span class="o">=</span> <span class="n">point0</span> <span class="o">+</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">-</span> <span class="n">point1</span><span class="p">)</span>
        <span class="n">len1</span> <span class="o">=</span> <span class="n">point1</span> <span class="o">+</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">-</span> <span class="n">point0</span><span class="p">)</span>
        <span class="n">new_genome0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">([</span><span class="n">len0</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">new_genome1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">([</span><span class="n">len1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>

        <span class="c1">#Operate crossover</span>
        <span class="n">new_genome0</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point0</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point0</span><span class="p">]</span>
        <span class="n">new_genome0</span><span class="p">[</span><span class="n">point0</span><span class="p">:]</span> <span class="o">=</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">point1</span><span class="p">:]</span>
        <span class="n">new_genome1</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point1</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point1</span><span class="p">]</span>
        <span class="n">new_genome1</span><span class="p">[</span><span class="n">point1</span><span class="p">:]</span> <span class="o">=</span> <span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">point0</span><span class="p">:]</span>

    <span class="k">elif</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
        <span class="c1">#Operate crossover</span>
        <span class="n">new_genome0</span> <span class="o">=</span> <span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point0</span><span class="p">]</span> <span class="o">+</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">point1</span><span class="p">:]</span>
        <span class="n">new_genome1</span> <span class="o">=</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point1</span><span class="p">]</span> <span class="o">+</span> <span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">point0</span><span class="p">:]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown genome representation&quot;</span><span class="p">)</span>

    <span class="n">new_ind0</span> <span class="o">=</span> <span class="n">reMap_leap</span><span class="p">(</span><span class="n">parent0</span><span class="p">,</span> <span class="n">new_genome0</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>
    <span class="n">new_ind1</span> <span class="o">=</span> <span class="n">reMap_leap</span><span class="p">(</span><span class="n">parent1</span><span class="p">,</span> <span class="n">new_genome1</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">new_ind0</span><span class="o">.</span><span class="n">depth</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
        <span class="n">new_ind0</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">if</span> <span class="n">new_ind1</span><span class="o">.</span><span class="n">depth</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
        <span class="n">new_ind1</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">max_genome_length</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_ind0</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_genome_length</span><span class="p">:</span>
            <span class="n">new_ind0</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_ind1</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_genome_length</span><span class="p">:</span>
            <span class="n">new_ind1</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">return</span> <span class="n">new_ind0</span><span class="p">,</span> <span class="n">new_ind1</span>   
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.leap_mutation_int_flip_per_codon" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">leap_mutation_int_flip_per_codon</span><span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">mut_probability</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">max_genome_length</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Mutation operator for LEAP GE genome. Each codon within the effective used codons
of the genome are checked</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>individual to mutate</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>mut_probability</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>chance for each codon to mutate</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum value for a codon</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum allowed length of genome when not None</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>new_ind</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new individual incorporating any mutations</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1842</span>
<span class="normal">1843</span>
<span class="normal">1844</span>
<span class="normal">1845</span>
<span class="normal">1846</span>
<span class="normal">1847</span>
<span class="normal">1848</span>
<span class="normal">1849</span>
<span class="normal">1850</span>
<span class="normal">1851</span>
<span class="normal">1852</span>
<span class="normal">1853</span>
<span class="normal">1854</span>
<span class="normal">1855</span>
<span class="normal">1856</span>
<span class="normal">1857</span>
<span class="normal">1858</span>
<span class="normal">1859</span>
<span class="normal">1860</span>
<span class="normal">1861</span>
<span class="normal">1862</span>
<span class="normal">1863</span>
<span class="normal">1864</span>
<span class="normal">1865</span>
<span class="normal">1866</span>
<span class="normal">1867</span>
<span class="normal">1868</span>
<span class="normal">1869</span>
<span class="normal">1870</span>
<span class="normal">1871</span>
<span class="normal">1872</span>
<span class="normal">1873</span>
<span class="normal">1874</span>
<span class="normal">1875</span>
<span class="normal">1876</span>
<span class="normal">1877</span>
<span class="normal">1878</span>
<span class="normal">1879</span>
<span class="normal">1880</span>
<span class="normal">1881</span>
<span class="normal">1882</span>
<span class="normal">1883</span>
<span class="normal">1884</span>
<span class="normal">1885</span>
<span class="normal">1886</span>
<span class="normal">1887</span>
<span class="normal">1888</span>
<span class="normal">1889</span>
<span class="normal">1890</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">leap_mutation_int_flip_per_codon</span><span class="p">(</span><span class="n">ind</span><span class="p">:</span><span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> <span class="n">mut_probability</span><span class="p">:</span><span class="nb">float</span><span class="p">,</span> 
                                <span class="n">codon_size</span><span class="p">:</span><span class="nb">int</span> <span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
                                <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
                                <span class="n">max_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Mutation operator for LEAP GE genome. Each codon within the effective used codons</span>
<span class="sd">    of the genome are checked</span>

<span class="sd">    Args:</span>
<span class="sd">        ind: individual to mutate</span>
<span class="sd">        mut_probability: chance for each codon to mutate</span>
<span class="sd">        codon_size: maximum value for a codon</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        max_genome_length: maximum allowed length of genome when not None</span>

<span class="sd">    Returns:</span>
<span class="sd">        new_ind: new individual incorporating any mutations</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="c1"># Operation mutation within the effective part of the genome</span>
    <span class="k">if</span> <span class="n">ind</span><span class="o">.</span><span class="n">invalid</span><span class="p">:</span> <span class="c1">#used_codons = 0</span>
        <span class="n">possible_mutation_codons</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">possible_mutation_codons</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">),</span> <span class="n">ind</span><span class="o">.</span><span class="n">used_codons</span><span class="p">)</span> <span class="c1">#in case of wrapping, used_codons can be greater than genome&#39;s length</span>

    <span class="n">mutated_</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">possible_mutation_codons</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">mut_probability</span><span class="p">:</span>
            <span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">)</span>
            <span class="n">mutated_</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">mutated_</span><span class="p">:</span>
        <span class="n">new_ind</span> <span class="o">=</span> <span class="n">reMap_leap</span><span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">new_ind</span> <span class="o">=</span> <span class="n">ind</span>


    <span class="k">if</span> <span class="n">new_ind</span><span class="o">.</span><span class="n">depth</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
        <span class="n">new_ind</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">max_genome_length</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_genome_length</span><span class="p">:</span>
            <span class="n">new_ind</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">mutated_</span><span class="p">:</span>
        <span class="k">del</span> <span class="n">new_ind</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span>
    <span class="k">return</span> <span class="n">new_ind</span><span class="p">,</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.leap_random_initialisation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">leap_random_initialisation</span><span class="p">(</span><span class="n">ind_class</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">min_init_genome_length</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">genome_representation</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Randomly generated linear genome compatible with LEAP chromosome mapping. 
The genomes contain complete frames</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind_class</code></td>
            <td>
                  <code><span title="deap.creator.MetaCreator">MetaCreator</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Deap creator</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>pop_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of individuals to create</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>min_init_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>minimum number of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum number of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth for tree in mapping genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum to be stored in codon</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genome_representation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>either list or numpy</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>population</code></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of individuals</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span>
<span class="normal">936</span>
<span class="normal">937</span>
<span class="normal">938</span>
<span class="normal">939</span>
<span class="normal">940</span>
<span class="normal">941</span>
<span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span>
<span class="normal">946</span>
<span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span>
<span class="normal">950</span>
<span class="normal">951</span>
<span class="normal">952</span>
<span class="normal">953</span>
<span class="normal">954</span>
<span class="normal">955</span>
<span class="normal">956</span>
<span class="normal">957</span>
<span class="normal">958</span>
<span class="normal">959</span>
<span class="normal">960</span>
<span class="normal">961</span>
<span class="normal">962</span>
<span class="normal">963</span>
<span class="normal">964</span>
<span class="normal">965</span>
<span class="normal">966</span>
<span class="normal">967</span>
<span class="normal">968</span>
<span class="normal">969</span>
<span class="normal">970</span>
<span class="normal">971</span>
<span class="normal">972</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">leap_random_initialisation</span><span class="p">(</span><span class="n">ind_class</span><span class="p">:</span> <span class="s1">&#39;deap.creator.MetaCreator&#39;</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
                          <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> 
                          <span class="n">min_init_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">max_init_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
                          <span class="n">genome_representation</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Randomly generated linear genome compatible with LEAP chromosome mapping. </span>
<span class="sd">    The genomes contain complete frames</span>

<span class="sd">    Args:</span>
<span class="sd">        ind_class: Deap creator</span>
<span class="sd">        pop_size: number of individuals to create</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        min_init_genome_length: minimum number of codons in genome</span>
<span class="sd">        max_init_genome_length: maximum number of codons in genome</span>
<span class="sd">        max_init_depth: maximum depth for tree in mapping genome</span>
<span class="sd">        codon_size: maximum to be stored in codon</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        genome_representation: either list or numpy</span>


<span class="sd">    Returns:</span>
<span class="sd">        population: list of individuals</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="n">population</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># generate a complete set of frames for Leap </span>
    <span class="n">frame_size</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span>
    <span class="n">minsize</span> <span class="o">=</span> <span class="n">min_init_genome_length</span> <span class="o">//</span> <span class="n">frame_size</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">maxsize</span> <span class="o">=</span> <span class="n">max_init_genome_length</span> <span class="o">//</span> <span class="n">frame_size</span> <span class="o">+</span> <span class="mi">1</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">pop_size</span><span class="p">):</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">init_genome_length</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">minsize</span><span class="p">,</span> <span class="n">maxsize</span><span class="p">)</span> <span class="o">*</span> <span class="n">codon_size</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">init_genome_length</span><span class="p">):</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">))</span>
        <span class="n">ind</span> <span class="o">=</span> <span class="n">ind_class</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>
        <span class="n">population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">elif</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;numpy&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">population</span><span class="p">:</span>
            <span class="n">ind</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unkonwn genome representation&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.leap_sensible_initialisation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">leap_sensible_initialisation</span><span class="p">(</span><span class="n">ind_class</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">min_init_genome_length</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">genome_representation</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Sensible initialization for LEAP genome. Half the generated individuals
are up to the depth of the max_init_depth and half are limited to that depth but may be
smaller</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind_class</code></td>
            <td>
                  <code><span title="deap.creator.MetaCreator">MetaCreator</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Deap creator</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>pop_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of individuals to create</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>min_init_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>minimum number of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum number of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth for tree in mapping genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum to be stored in codon</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genome_representation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>either list or numpy</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>population</code></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of individuals</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span>
<span class="normal">1246</span>
<span class="normal">1247</span>
<span class="normal">1248</span>
<span class="normal">1249</span>
<span class="normal">1250</span>
<span class="normal">1251</span>
<span class="normal">1252</span>
<span class="normal">1253</span>
<span class="normal">1254</span>
<span class="normal">1255</span>
<span class="normal">1256</span>
<span class="normal">1257</span>
<span class="normal">1258</span>
<span class="normal">1259</span>
<span class="normal">1260</span>
<span class="normal">1261</span>
<span class="normal">1262</span>
<span class="normal">1263</span>
<span class="normal">1264</span>
<span class="normal">1265</span>
<span class="normal">1266</span>
<span class="normal">1267</span>
<span class="normal">1268</span>
<span class="normal">1269</span>
<span class="normal">1270</span>
<span class="normal">1271</span>
<span class="normal">1272</span>
<span class="normal">1273</span>
<span class="normal">1274</span>
<span class="normal">1275</span>
<span class="normal">1276</span>
<span class="normal">1277</span>
<span class="normal">1278</span>
<span class="normal">1279</span>
<span class="normal">1280</span>
<span class="normal">1281</span>
<span class="normal">1282</span>
<span class="normal">1283</span>
<span class="normal">1284</span>
<span class="normal">1285</span>
<span class="normal">1286</span>
<span class="normal">1287</span>
<span class="normal">1288</span>
<span class="normal">1289</span>
<span class="normal">1290</span>
<span class="normal">1291</span>
<span class="normal">1292</span>
<span class="normal">1293</span>
<span class="normal">1294</span>
<span class="normal">1295</span>
<span class="normal">1296</span>
<span class="normal">1297</span>
<span class="normal">1298</span>
<span class="normal">1299</span>
<span class="normal">1300</span>
<span class="normal">1301</span>
<span class="normal">1302</span>
<span class="normal">1303</span>
<span class="normal">1304</span>
<span class="normal">1305</span>
<span class="normal">1306</span>
<span class="normal">1307</span>
<span class="normal">1308</span>
<span class="normal">1309</span>
<span class="normal">1310</span>
<span class="normal">1311</span>
<span class="normal">1312</span>
<span class="normal">1313</span>
<span class="normal">1314</span>
<span class="normal">1315</span>
<span class="normal">1316</span>
<span class="normal">1317</span>
<span class="normal">1318</span>
<span class="normal">1319</span>
<span class="normal">1320</span>
<span class="normal">1321</span>
<span class="normal">1322</span>
<span class="normal">1323</span>
<span class="normal">1324</span>
<span class="normal">1325</span>
<span class="normal">1326</span>
<span class="normal">1327</span>
<span class="normal">1328</span>
<span class="normal">1329</span>
<span class="normal">1330</span>
<span class="normal">1331</span>
<span class="normal">1332</span>
<span class="normal">1333</span>
<span class="normal">1334</span>
<span class="normal">1335</span>
<span class="normal">1336</span>
<span class="normal">1337</span>
<span class="normal">1338</span>
<span class="normal">1339</span>
<span class="normal">1340</span>
<span class="normal">1341</span>
<span class="normal">1342</span>
<span class="normal">1343</span>
<span class="normal">1344</span>
<span class="normal">1345</span>
<span class="normal">1346</span>
<span class="normal">1347</span>
<span class="normal">1348</span>
<span class="normal">1349</span>
<span class="normal">1350</span>
<span class="normal">1351</span>
<span class="normal">1352</span>
<span class="normal">1353</span>
<span class="normal">1354</span>
<span class="normal">1355</span>
<span class="normal">1356</span>
<span class="normal">1357</span>
<span class="normal">1358</span>
<span class="normal">1359</span>
<span class="normal">1360</span>
<span class="normal">1361</span>
<span class="normal">1362</span>
<span class="normal">1363</span>
<span class="normal">1364</span>
<span class="normal">1365</span>
<span class="normal">1366</span>
<span class="normal">1367</span>
<span class="normal">1368</span>
<span class="normal">1369</span>
<span class="normal">1370</span>
<span class="normal">1371</span>
<span class="normal">1372</span>
<span class="normal">1373</span>
<span class="normal">1374</span>
<span class="normal">1375</span>
<span class="normal">1376</span>
<span class="normal">1377</span>
<span class="normal">1378</span>
<span class="normal">1379</span>
<span class="normal">1380</span>
<span class="normal">1381</span>
<span class="normal">1382</span>
<span class="normal">1383</span>
<span class="normal">1384</span>
<span class="normal">1385</span>
<span class="normal">1386</span>
<span class="normal">1387</span>
<span class="normal">1388</span>
<span class="normal">1389</span>
<span class="normal">1390</span>
<span class="normal">1391</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">leap_sensible_initialisation</span><span class="p">(</span><span class="n">ind_class</span><span class="p">:</span> <span class="s1">&#39;deap.creator.MetaCreator&#39;</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
                          <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> 
                          <span class="n">min_init_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">max_init_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
                          <span class="n">genome_representation</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Sensible initialization for LEAP genome. Half the generated individuals</span>
<span class="sd">    are up to the depth of the max_init_depth and half are limited to that depth but may be</span>
<span class="sd">    smaller</span>

<span class="sd">    Args:</span>
<span class="sd">        ind_class: Deap creator</span>
<span class="sd">        pop_size: number of individuals to create</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        min_init_genome_length: minimum number of codons in genome</span>
<span class="sd">        max_init_genome_length: maximum number of codons in genome</span>
<span class="sd">        max_init_depth: maximum depth for tree in mapping genome</span>
<span class="sd">        codon_size: maximum to be stored in codon</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        genome_representation: either list or numpy</span>


<span class="sd">    Returns:</span>
<span class="sd">        population: list of individuals</span>
<span class="sd">    &quot;&quot;&quot;</span> 
    <span class="c1">#Calculate the number of individuals to be generated with each method</span>
    <span class="n">is_odd</span> <span class="o">=</span> <span class="n">pop_size</span> <span class="o">%</span> <span class="mi">2</span>
    <span class="n">n_grow</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pop_size</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>

    <span class="n">n_sets_grow</span> <span class="o">=</span> <span class="n">max_init_depth</span> <span class="o">-</span> <span class="n">min_init_depth</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">set_size</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_grow</span><span class="o">/</span><span class="n">n_sets_grow</span><span class="p">)</span>
    <span class="n">remaining</span> <span class="o">=</span> <span class="n">n_grow</span> <span class="o">%</span> <span class="n">n_sets_grow</span>

    <span class="n">n_full</span> <span class="o">=</span> <span class="n">n_grow</span> <span class="o">+</span> <span class="n">is_odd</span> <span class="o">+</span> <span class="n">remaining</span> <span class="c1">#if pop_size is odd, generate an extra ind with &quot;full&quot;</span>

    <span class="c1">#TODO check if it is possible to generate inds with max_init_depth        </span>

    <span class="n">population</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1">#Generate inds using &quot;Grow&quot;</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_sets_grow</span><span class="p">):</span>
        <span class="n">max_init_depth_</span> <span class="o">=</span> <span class="n">min_init_depth</span> <span class="o">+</span> <span class="n">i</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">set_size</span><span class="p">):</span>
            <span class="n">remainders</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the choices</span>
            <span class="n">possible_choices</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the respective possible choices</span>
            <span class="n">rules_used</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#registers which rule used for setting the frame</span>

            <span class="n">phenotype</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">start_rule</span>
            <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span> <span class="c1">#</span>
            <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="c1">#it keeps the depth of each branch</span>
            <span class="n">idx_branch</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1">#index of the current branch being grown</span>
            <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">idx_NT</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">total_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]]</span>
                <span class="n">actual_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]</span> <span class="k">if</span> <span class="n">PR</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_init_depth_</span><span class="p">]</span>
                <span class="n">Ch</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">actual_options</span><span class="p">)</span>
                <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
                <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
                    <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                    <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>
                    <span class="n">rules_used</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx_NT</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                        <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>
                        <span class="n">rules_used</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx_NT</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">idx_branch</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">depths</span> <span class="o">=</span> <span class="n">depths</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
                <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;terminal&#39;</span><span class="p">:</span>
                    <span class="n">idx_branch</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span>

            <span class="c1">#Generate the genome</span>
            <span class="c1">#remainders contain the choice for the rule</span>
            <span class="c1">#possible_choices simply tells you how many choices for that codon -- needed to generate actual codon value</span>
            <span class="c1"># self.rule_conversion.append(idx)</span>
            <span class="c1"># self.nt_rule_size += 1</span>
            <span class="n">genome</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span> <span class="ow">or</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
                <span class="n">frame</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">remainders</span><span class="p">)):</span>
                    <span class="n">codon</span> <span class="o">=</span> <span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">int</span><span class="p">(</span><span class="mf">1e10</span><span class="p">))</span> <span class="o">%</span> <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(((</span><span class="n">codon_size</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span> <span class="o">*</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">k</span><span class="p">])</span> <span class="o">+</span> <span class="n">remainders</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
                    <span class="c1"># check if frame has been used</span>
                    <span class="k">if</span> <span class="n">frame</span><span class="p">[</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">[</span><span class="n">rules_used</span><span class="p">[</span><span class="n">k</span><span class="p">]]]</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                        <span class="c1"># fill unused codons in frame randomly</span>
                        <span class="n">frame</span> <span class="o">=</span> <span class="p">[</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">codon_size</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">frame</span><span class="p">]</span>
                        <span class="n">genome</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">frame</span><span class="p">)</span>
                        <span class="n">frame</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>
                    <span class="c1"># TODO</span>
                    <span class="n">frame</span><span class="p">[</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">[</span><span class="n">rules_used</span><span class="p">[</span><span class="n">k</span><span class="p">]]]</span><span class="o">=</span><span class="n">codon</span>
                <span class="k">if</span> <span class="n">frame</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">frame</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">frame</span><span class="p">)</span> <span class="ow">or</span> <span class="n">frame</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                    <span class="n">frame</span> <span class="o">=</span> <span class="p">[</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">codon_size</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">frame</span><span class="p">]</span>
                    <span class="n">genome</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">frame</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>

            <span class="c1">#Include a tail with 50% of the genome&#39;s size</span>
            <span class="c1"># calculate based on number of frames and extend by complete frames</span>
            <span class="n">n_frames</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="o">/</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span>
            <span class="n">size_tail</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.5</span><span class="o">*</span><span class="n">n_frames</span><span class="p">)</span><span class="o">*</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">),</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">))</span> <span class="c1">#Tail must have at least one codon. Otherwise, in the lazy approach, when we have the last PR with just a single option, the mapping procces will not terminate.</span>

            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">size_tail</span><span class="p">):</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">codon_size</span><span class="p">))</span>

            <span class="c1">#Initialise the individual and include in the population</span>
            <span class="n">ind</span> <span class="o">=</span> <span class="n">ind_class</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_init_depth_</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>

            <span class="c1">#Check if the individual was mapped correctly</span>
            <span class="k">if</span> <span class="n">remainders</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">structure</span> <span class="ow">or</span> <span class="n">phenotype</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">phenotype</span> <span class="ow">or</span> <span class="nb">max</span><span class="p">(</span><span class="n">depths</span><span class="p">)</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">depth</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;error in the mapping&#39;</span><span class="p">)</span>
            <span class="n">population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>    

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_full</span><span class="p">):</span>
        <span class="n">remainders</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the choices</span>
        <span class="n">possible_choices</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the respective possible choices</span>
        <span class="n">rules_used</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#registers which rule used for setting the frame</span>

        <span class="n">phenotype</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">start_rule</span>
        <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span> <span class="c1">#</span>
        <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="c1">#it keeps the depth of each branch</span>
        <span class="n">idx_branch</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1">#index of the current branch being grown</span>

        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">idx_NT</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">total_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]]</span>
            <span class="n">actual_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]</span> <span class="k">if</span> <span class="n">PR</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_init_depth</span><span class="p">]</span>
            <span class="n">recursive_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">actual_options</span> <span class="k">if</span> <span class="n">PR</span><span class="p">[</span><span class="mi">4</span><span class="p">]]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">recursive_options</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">Ch</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">recursive_options</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">Ch</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">actual_options</span><span class="p">)</span>
            <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
                <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>
                <span class="n">rules_used</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx_NT</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                    <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>
                    <span class="n">rules_used</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx_NT</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">idx_branch</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">depths</span> <span class="o">=</span> <span class="n">depths</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;terminal&#39;</span><span class="p">:</span>
                <span class="n">idx_branch</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span>

        <span class="c1">#Generate the genome</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span> <span class="ow">or</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
            <span class="n">frame</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">remainders</span><span class="p">)):</span>
                <span class="n">codon</span> <span class="o">=</span> <span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">int</span><span class="p">(</span><span class="mf">1e10</span><span class="p">))</span> <span class="o">%</span> <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(((</span><span class="n">codon_size</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">j</span><span class="p">]))</span> <span class="o">*</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">j</span><span class="p">])</span> <span class="o">+</span> <span class="n">remainders</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">frame</span><span class="p">[</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">[</span><span class="n">rules_used</span><span class="p">[</span><span class="n">j</span><span class="p">]]]</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                    <span class="c1"># fill unused codons in frame randomly</span>
                    <span class="n">frame</span> <span class="o">=</span> <span class="p">[</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">codon_size</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">frame</span><span class="p">]</span>
                    <span class="n">genome</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">frame</span><span class="p">)</span>
                    <span class="n">frame</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>
                <span class="n">frame</span><span class="p">[</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">[</span><span class="n">rules_used</span><span class="p">[</span><span class="n">j</span><span class="p">]]]</span><span class="o">=</span><span class="n">codon</span>
            <span class="k">if</span> <span class="n">frame</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">frame</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">frame</span><span class="p">)</span> <span class="ow">or</span> <span class="n">frame</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="n">frame</span> <span class="o">=</span> <span class="p">[</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">codon_size</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span> <span class="k">else</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">frame</span><span class="p">]</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">frame</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>

        <span class="c1">#Include a tail with 50% of the genome&#39;s size</span>
        <span class="c1"># calculate based on number of frames and extend by complete frames</span>
        <span class="n">n_frames</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="o">/</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span>
        <span class="n">size_tail</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.5</span><span class="o">*</span><span class="n">n_frames</span><span class="p">)</span><span class="o">*</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">),</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">))</span> <span class="c1">#Tail must have at least one codon. Otherwise, in the lazy approach, when we have the last PR with just a single option, the mapping procces will not terminate.</span>

        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">size_tail</span><span class="p">):</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">codon_size</span><span class="p">))</span>

        <span class="c1">#Initialise the individual and include in the population</span>
        <span class="n">ind</span> <span class="o">=</span> <span class="n">ind_class</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>

        <span class="c1">#Check if the individual was mapped correctly</span>
        <span class="k">if</span> <span class="n">remainders</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">structure</span> <span class="ow">or</span> <span class="n">phenotype</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">phenotype</span> <span class="ow">or</span> <span class="nb">max</span><span class="p">(</span><span class="n">depths</span><span class="p">)</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">depth</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;error in the mapping&#39;</span><span class="p">)</span>

        <span class="n">population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>    

    <span class="k">if</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">elif</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;numpy&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">population</span><span class="p">:</span>
            <span class="n">ind</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unkonwn genome representation&quot;</span><span class="p">)</span>    
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mapper_eager" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mapper_eager</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Maps standard GE single chromosome genome. Uses eager mapping so that 
every nonterminal in the grammar consumes a codon in the genome</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>genome</code></td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.grammar">grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar used to create individual phenotype</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>phenotype</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>nodes</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of nodes in tree</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>depth</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>max depth of tree used in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>used_codons</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of codons consumed in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>invalid</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if valid phenotype not constructed</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int for number of wraps used (0 for this implementation as wrapping not used)</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>structure</code></td>            <td>
                  <code>tuple[str, int, int, int, bool, list]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of production rules chosen in mapping, 
used to monitor population diversity</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mapper_eager</span><span class="p">(</span><span class="n">genome</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">grammar</span><span class="p">:</span> <span class="s2">&quot;grape.grape.grammar&quot;</span><span class="p">,</span>
                  <span class="n">max_depth</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">bool</span><span class="p">,</span> <span class="nb">list</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Maps standard GE single chromosome genome. Uses eager mapping so that </span>
<span class="sd">    every nonterminal in the grammar consumes a codon in the genome</span>

<span class="sd">    Args:</span>
<span class="sd">        genome: list of codons in genome</span>
<span class="sd">        grammar: BNF grammar used to create individual phenotype</span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>


<span class="sd">    Returns:</span>
<span class="sd">        phenotype: </span>
<span class="sd">        nodes: number of nodes in tree</span>
<span class="sd">        depth: max depth of tree used in mapping</span>
<span class="sd">        used_codons: number of codons consumed in mapping</span>
<span class="sd">        invalid: True if valid phenotype not constructed</span>
<span class="sd">        int for number of wraps used (0 for this implementation as wrapping not used)</span>
<span class="sd">        structure: list of production rules chosen in mapping, </span>
<span class="sd">            used to monitor population diversity</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">idx_genome</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">phenotype</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">start_rule</span>
    <span class="n">next_NT</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
    <span class="n">n_starting_NTs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">term</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)])</span>
    <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="n">n_starting_NTs</span> <span class="c1">#it keeps the depth of each branch</span>
    <span class="n">idx_depth</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nodes</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">structure</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">while</span> <span class="n">next_NT</span> <span class="ow">and</span> <span class="n">idx_genome</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">):</span>
        <span class="n">NT_index</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">next_NT</span><span class="p">)</span>
        <span class="n">index_production_chosen</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="n">idx_genome</span><span class="p">]</span> <span class="o">%</span> <span class="n">grammar</span><span class="o">.</span><span class="n">n_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span>
        <span class="n">structure</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">index_production_chosen</span><span class="p">)</span>
        <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">next_NT</span><span class="p">,</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="k">if</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="c1">#arity 0 (T)</span>
            <span class="n">idx_depth</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">nodes</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#arity 1 (PR with one NT)</span>
            <span class="k">pass</span>        
        <span class="k">else</span><span class="p">:</span> <span class="c1">#it is a PR with more than one NT</span>
            <span class="n">arity</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">idx_depth</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="n">list_depth</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>

        <span class="n">next_</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">next_</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="n">next_</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">idx_genome</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">if</span> <span class="n">next_NT</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="n">idx_genome</span>

    <span class="n">depth</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">list_depth</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">phenotype</span><span class="p">,</span> <span class="n">nodes</span><span class="p">,</span> <span class="n">depth</span><span class="p">,</span> <span class="n">used_codons</span><span class="p">,</span> <span class="n">invalid</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">structure</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mapper_eager_leap" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mapper_eager_leap</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Maps GE LEAP chromosome genome. Uses eager mapping so that 
every nonterminal in the grammar consumes a codon in the genome</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>genome</code></td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.grammar">grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar used to create individual phenotype</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>phenotype</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>nodes</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of nodes in tree</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>depth</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>max depth of tree used in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>used_codons</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of frames consumed in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>invalid</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if valid phenotype not constructed</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int for number of wraps used (0 for this implementation as wrapping not used)</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>structure</code></td>            <td>
                  <code>tuple[str, int, int, int, bool, list]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of production rules chosen in mapping, 
used to monitor population diversity</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mapper_eager_leap</span><span class="p">(</span><span class="n">genome</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">grammar</span><span class="p">:</span> <span class="s2">&quot;grape.grape.grammar&quot;</span><span class="p">,</span>
                  <span class="n">max_depth</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">bool</span><span class="p">,</span> <span class="nb">list</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Maps GE LEAP chromosome genome. Uses eager mapping so that </span>
<span class="sd">    every nonterminal in the grammar consumes a codon in the genome</span>

<span class="sd">    Args:</span>
<span class="sd">        genome: list of codons in genome</span>
<span class="sd">        grammar: BNF grammar used to create individual phenotype</span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>


<span class="sd">    Returns:</span>
<span class="sd">        phenotype: </span>
<span class="sd">        nodes: number of nodes in tree</span>
<span class="sd">        depth: max depth of tree used in mapping</span>
<span class="sd">        used_codons: number of frames consumed in mapping</span>
<span class="sd">        invalid: True if valid phenotype not constructed</span>
<span class="sd">        int for number of wraps used (0 for this implementation as wrapping not used)</span>
<span class="sd">        structure: list of production rules chosen in mapping, </span>
<span class="sd">            used to monitor population diversity</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">phenotype</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">start_rule</span>
    <span class="n">next_NT</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
    <span class="n">n_starting_NTs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">term</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)])</span>
    <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="n">n_starting_NTs</span> <span class="c1">#it keeps the depth of each branch</span>
    <span class="n">idx_depth</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nodes</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">structure</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># move along the genome in frames</span>
    <span class="c1"># consume codons in a frame until you find a repeat</span>
    <span class="c1"># shift to new frame </span>
    <span class="n">idx_frame</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">consumed_codons</span> <span class="o">=</span> <span class="p">[</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>
    <span class="n">n_frames</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="o">/</span> <span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span>

    <span class="k">while</span> <span class="n">next_NT</span> <span class="ow">and</span> <span class="n">idx_frame</span> <span class="o">&lt;</span> <span class="n">n_frames</span><span class="p">:</span>
        <span class="n">NT_index</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">next_NT</span><span class="p">)</span>
        <span class="n">codon_frame_idx</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span>
        <span class="k">if</span><span class="p">(</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">codon_frame_idx</span><span class="p">]):</span>
            <span class="n">idx_frame</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">idx_frame</span> <span class="o">==</span> <span class="n">n_frames</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="n">consumed_codons</span> <span class="o">=</span> <span class="p">[</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>

        <span class="n">index_production_chosen</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="n">idx_frame</span><span class="o">*</span><span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="o">+</span><span class="n">codon_frame_idx</span><span class="p">]</span> <span class="o">%</span> <span class="n">grammar</span><span class="o">.</span><span class="n">n_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span>
        <span class="n">structure</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">index_production_chosen</span><span class="p">)</span>
        <span class="n">consumed_codons</span><span class="p">[</span><span class="n">codon_frame_idx</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">next_NT</span><span class="p">,</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="k">if</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="c1">#arity 0 (T)</span>
            <span class="n">idx_depth</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">nodes</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#arity 1 (PR with one NT)</span>
            <span class="k">pass</span>        
        <span class="k">else</span><span class="p">:</span> <span class="c1">#it is a PR with more than one NT</span>
            <span class="n">arity</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">idx_depth</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="n">list_depth</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>

        <span class="n">next_</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">next_</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="n">next_</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">next_NT</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">depth</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">list_depth</span><span class="p">)</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="n">idx_frame</span> <span class="o">*</span> <span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span>

    <span class="n">depth</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">list_depth</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">phenotype</span><span class="p">,</span> <span class="n">nodes</span><span class="p">,</span> <span class="n">depth</span><span class="p">,</span> <span class="n">used_codons</span><span class="p">,</span> <span class="n">invalid</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">structure</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mapper_eager_mcge" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mapper_eager_mcge</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Maps GE multi-chromosome genome (MCGE). Uses eager mapping so that 
every nonterminal in the grammar consumes a codon in the genome</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>genome</code></td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.grammar">grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar used to create individual phenotype</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>phenotype</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>nodes</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of nodes in tree</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>depth</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>max depth of tree used in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>used_codons</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of codons consumed in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>invalid</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if valid phenotype not constructed</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int for number of wraps used (0 for this implementation as wrapping not used)</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>structure</code></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of production rules chosen in mapping, 
used to monitor population diversity</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>consumed_codons</code></td>            <td>
                  <code>tuple[str, int, int, int, bool, list, list]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>2D list containing max codon used in each chromosome</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mapper_eager_mcge</span><span class="p">(</span><span class="n">genome</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">grammar</span><span class="p">:</span> <span class="s2">&quot;grape.grape.grammar&quot;</span><span class="p">,</span>
                  <span class="n">max_depth</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">bool</span><span class="p">,</span> <span class="nb">list</span><span class="p">,</span> <span class="nb">list</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Maps GE multi-chromosome genome (MCGE). Uses eager mapping so that </span>
<span class="sd">    every nonterminal in the grammar consumes a codon in the genome</span>

<span class="sd">    Args:</span>
<span class="sd">        genome: list of codons in genome</span>
<span class="sd">        grammar: BNF grammar used to create individual phenotype</span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>


<span class="sd">    Returns:</span>
<span class="sd">        phenotype: </span>
<span class="sd">        nodes: number of nodes in tree</span>
<span class="sd">        depth: max depth of tree used in mapping</span>
<span class="sd">        used_codons: number of codons consumed in mapping</span>
<span class="sd">        invalid: True if valid phenotype not constructed</span>
<span class="sd">        int for number of wraps used (0 for this implementation as wrapping not used)</span>
<span class="sd">        structure: list of production rules chosen in mapping, </span>
<span class="sd">            used to monitor population diversity</span>
<span class="sd">        consumed_codons: 2D list containing max codon used in each chromosome</span>
<span class="sd">    &quot;&quot;&quot;</span>  

    <span class="n">phenotype</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">start_rule</span>
    <span class="n">next_NT</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
    <span class="n">n_starting_NTs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">term</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)])</span>
    <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="n">n_starting_NTs</span> <span class="c1">#it keeps the depth of each branch</span>
    <span class="n">idx_depth</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nodes</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">structure</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># track position in each chromosome</span>
    <span class="n">consumed_codons</span> <span class="o">=</span> <span class="p">[</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>

    <span class="k">while</span> <span class="n">next_NT</span><span class="p">:</span>
        <span class="n">NT_index</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">next_NT</span><span class="p">)</span>
        <span class="n">chr_idx</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span>
        <span class="k">if</span><span class="p">(</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">chr_idx</span><span class="p">]</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="n">chr_idx</span><span class="p">])):</span>
            <span class="k">break</span>

        <span class="n">index_production_chosen</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="n">chr_idx</span><span class="p">][</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">chr_idx</span><span class="p">]]</span> <span class="o">%</span> <span class="n">grammar</span><span class="o">.</span><span class="n">n_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span>
        <span class="n">consumed_codons</span><span class="p">[</span><span class="n">chr_idx</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">structure</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">index_production_chosen</span><span class="p">)</span>

        <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">next_NT</span><span class="p">,</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="k">if</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="c1">#arity 0 (T)</span>
            <span class="n">idx_depth</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">nodes</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#arity 1 (PR with one NT)</span>
            <span class="k">pass</span>        
        <span class="k">else</span><span class="p">:</span> <span class="c1">#it is a PR with more than one NT</span>
            <span class="n">arity</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">idx_depth</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="n">list_depth</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>

        <span class="n">next_</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">next_</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="n">next_</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">next_NT</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">consumed_codons</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">consumed_codons</span><span class="p">)</span>

    <span class="n">depth</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">list_depth</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">phenotype</span><span class="p">,</span> <span class="n">nodes</span><span class="p">,</span> <span class="n">depth</span><span class="p">,</span> <span class="n">used_codons</span><span class="p">,</span> <span class="n">invalid</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">structure</span><span class="p">,</span> <span class="n">consumed_codons</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mapper_lazy" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mapper_lazy</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Maps GE multi-chromosome genome (MCGE). Uses lazy mapping so that 
only nonterminals with more than one option in the grammar 
consume a codon in the genome</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>genome</code></td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.grammar">grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar used to create individual phenotype</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>phenotype</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>nodes</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of nodes in tree</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>depth</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>max depth of tree used in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>used_codons</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of codons consumed in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>invalid</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if valid phenotype not constructed</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int for number of wraps used (0 for this implementation as wrapping not used)</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>structure</code></td>            <td>
                  <code>tuple[str, int, int, int, bool, list]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of production rules chosen in mapping, 
used to monitor population diversity</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mapper_lazy</span><span class="p">(</span><span class="n">genome</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">grammar</span><span class="p">:</span> <span class="s2">&quot;grape.grape.grammar&quot;</span><span class="p">,</span>
                  <span class="n">max_depth</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">bool</span><span class="p">,</span> <span class="nb">list</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Maps GE multi-chromosome genome (MCGE). Uses lazy mapping so that </span>
<span class="sd">    only nonterminals with more than one option in the grammar </span>
<span class="sd">    consume a codon in the genome</span>

<span class="sd">    Args:</span>
<span class="sd">        genome: list of codons in genome</span>
<span class="sd">        grammar: BNF grammar used to create individual phenotype</span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>


<span class="sd">    Returns:</span>
<span class="sd">        phenotype: </span>
<span class="sd">        nodes: number of nodes in tree</span>
<span class="sd">        depth: max depth of tree used in mapping</span>
<span class="sd">        used_codons: number of codons consumed in mapping</span>
<span class="sd">        invalid: True if valid phenotype not constructed</span>
<span class="sd">        int for number of wraps used (0 for this implementation as wrapping not used)</span>
<span class="sd">        structure: list of production rules chosen in mapping, </span>
<span class="sd">            used to monitor population diversity</span>
<span class="sd">    &quot;&quot;&quot;</span>  

    <span class="n">idx_genome</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">phenotype</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">start_rule</span>
    <span class="n">next_NT</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
    <span class="n">n_starting_NTs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">term</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)])</span>
    <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="n">n_starting_NTs</span> <span class="c1">#it keeps the depth of each branch</span>
    <span class="n">idx_depth</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nodes</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">structure</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">while</span> <span class="n">next_NT</span> <span class="ow">and</span> <span class="n">idx_genome</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">):</span>
        <span class="n">NT_index</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">next_NT</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">grammar</span><span class="o">.</span><span class="n">n_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#there is a single PR for this non-terminal</span>
            <span class="n">index_production_chosen</span> <span class="o">=</span> <span class="mi">0</span>        
        <span class="k">else</span><span class="p">:</span> <span class="c1">#we consume one codon, and add the index to the structure</span>
            <span class="n">index_production_chosen</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="n">idx_genome</span><span class="p">]</span> <span class="o">%</span> <span class="n">grammar</span><span class="o">.</span><span class="n">n_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span>
            <span class="n">structure</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">index_production_chosen</span><span class="p">)</span>
            <span class="n">idx_genome</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">next_NT</span><span class="p">,</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="k">if</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="c1">#arity 0 (T)</span>
            <span class="n">idx_depth</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">nodes</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#arity 1 (PR with one NT)</span>
            <span class="k">pass</span>        
        <span class="k">else</span><span class="p">:</span> <span class="c1">#it is a PR with more than one NT</span>
            <span class="n">arity</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">idx_depth</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="n">list_depth</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>

        <span class="n">next_</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">next_</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="n">next_</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="kc">None</span>


    <span class="k">if</span> <span class="n">next_NT</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="n">idx_genome</span>

    <span class="n">depth</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">list_depth</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">phenotype</span><span class="p">,</span> <span class="n">nodes</span><span class="p">,</span> <span class="n">depth</span><span class="p">,</span> <span class="n">used_codons</span><span class="p">,</span> <span class="n">invalid</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">structure</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mapper_lazy_leap" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mapper_lazy_leap</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Maps GE multi-chromosome genome (MCGE). Uses lazy mapping so that 
only nonterminals with more than one option in the grammar 
consume a codon in the genome</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>genome</code></td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.grammar">grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar used to create individual phenotype</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>phenotype</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>nodes</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of nodes in tree</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>depth</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>max depth of tree used in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>used_codons</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of frames consumed in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>invalid</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if valid phenotype not constructed</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int for number of wraps used (0 for this implementation as wrapping not used)</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>structure</code></td>            <td>
                  <code>tuple[str, int, int, int, bool, list]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of production rules chosen in mapping, 
used to monitor population diversity</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mapper_lazy_leap</span><span class="p">(</span><span class="n">genome</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">grammar</span><span class="p">:</span> <span class="s2">&quot;grape.grape.grammar&quot;</span><span class="p">,</span>
                  <span class="n">max_depth</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">bool</span><span class="p">,</span> <span class="nb">list</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Maps GE multi-chromosome genome (MCGE). Uses lazy mapping so that </span>
<span class="sd">    only nonterminals with more than one option in the grammar </span>
<span class="sd">    consume a codon in the genome</span>

<span class="sd">    Args:</span>
<span class="sd">        genome: list of codons in genome</span>
<span class="sd">        grammar: BNF grammar used to create individual phenotype</span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>


<span class="sd">    Returns:</span>
<span class="sd">        phenotype: </span>
<span class="sd">        nodes: number of nodes in tree</span>
<span class="sd">        depth: max depth of tree used in mapping</span>
<span class="sd">        used_codons: number of frames consumed in mapping</span>
<span class="sd">        invalid: True if valid phenotype not constructed</span>
<span class="sd">        int for number of wraps used (0 for this implementation as wrapping not used)</span>
<span class="sd">        structure: list of production rules chosen in mapping, </span>
<span class="sd">            used to monitor population diversity</span>
<span class="sd">    &quot;&quot;&quot;</span>  

    <span class="n">phenotype</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">start_rule</span>
    <span class="n">next_NT</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
    <span class="n">n_starting_NTs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">term</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)])</span>
    <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="n">n_starting_NTs</span> <span class="c1">#it keeps the depth of each branch</span>
    <span class="n">idx_depth</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nodes</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">structure</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># move along the genome in frames</span>
    <span class="c1"># consume codons in a frame until you find a repeat</span>
    <span class="c1"># shift to new frame </span>
    <span class="n">idx_frame</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">consumed_codons</span> <span class="o">=</span> <span class="p">[</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>
    <span class="n">n_frames</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="o">/</span> <span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span>

    <span class="k">while</span> <span class="n">next_NT</span> <span class="ow">and</span> <span class="n">idx_frame</span> <span class="o">&lt;</span> <span class="n">n_frames</span><span class="p">:</span>
        <span class="n">NT_index</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">next_NT</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">grammar</span><span class="o">.</span><span class="n">n_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#there is a single PR for this non-terminal</span>
            <span class="n">index_production_chosen</span> <span class="o">=</span> <span class="mi">0</span>        
        <span class="k">else</span><span class="p">:</span> <span class="c1">#we consume one codon, and add the index to the structure</span>
            <span class="c1"># check if already used codon in this frame</span>
            <span class="n">codon_frame_idx</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span>
            <span class="k">if</span><span class="p">(</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">codon_frame_idx</span><span class="p">]):</span>
                <span class="n">idx_frame</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">idx_frame</span> <span class="o">==</span> <span class="n">n_frames</span><span class="p">:</span>
                    <span class="k">break</span>
                <span class="n">consumed_codons</span> <span class="o">=</span> <span class="p">[</span><span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>
            <span class="n">index_production_chosen</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="n">idx_frame</span><span class="o">*</span><span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="o">+</span><span class="n">codon_frame_idx</span><span class="p">]</span> <span class="o">%</span> <span class="n">grammar</span><span class="o">.</span><span class="n">n_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span>
            <span class="n">structure</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">index_production_chosen</span><span class="p">)</span>
            <span class="n">consumed_codons</span><span class="p">[</span><span class="n">codon_frame_idx</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">next_NT</span><span class="p">,</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="k">if</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="c1">#arity 0 (T)</span>
            <span class="n">idx_depth</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">nodes</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#arity 1 (PR with one NT)</span>
            <span class="k">pass</span>        
        <span class="k">else</span><span class="p">:</span> <span class="c1">#it is a PR with more than one NT</span>
            <span class="n">arity</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">idx_depth</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="n">list_depth</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>

        <span class="n">next_</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">next_</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="n">next_</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="kc">None</span>


    <span class="k">if</span> <span class="n">next_NT</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="n">idx_frame</span> <span class="o">*</span> <span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span>

    <span class="n">depth</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">list_depth</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">phenotype</span><span class="p">,</span> <span class="n">nodes</span><span class="p">,</span> <span class="n">depth</span><span class="p">,</span> <span class="n">used_codons</span><span class="p">,</span> <span class="n">invalid</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">structure</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mapper_lazy_mcge" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mapper_lazy_mcge</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Maps GE multi-chromosome genome (MCGE). Uses lazy mapping so that 
only nonterminals with more than one option in the grammar 
consume a codon in the genome</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>genome</code></td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.grammar">grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar used to create individual phenotype</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>phenotype</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>nodes</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of nodes in tree</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>depth</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>max depth of tree used in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>used_codons</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of codons consumed in mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>invalid</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if valid phenotype not constructed</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int for number of wraps used (0 for this implementation as wrapping not used)</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>structure</code></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of production rules chosen in mapping, 
used to monitor population diversity</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>consumed_codons</code></td>            <td>
                  <code>tuple[str, int, int, int, bool, list, list]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>2D list containing max codon used in each chromosome</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mapper_lazy_mcge</span><span class="p">(</span><span class="n">genome</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">grammar</span><span class="p">:</span> <span class="s2">&quot;grape.grape.grammar&quot;</span><span class="p">,</span>
                  <span class="n">max_depth</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">bool</span><span class="p">,</span> <span class="nb">list</span><span class="p">,</span> <span class="nb">list</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Maps GE multi-chromosome genome (MCGE). Uses lazy mapping so that </span>
<span class="sd">    only nonterminals with more than one option in the grammar </span>
<span class="sd">    consume a codon in the genome</span>

<span class="sd">    Args:</span>
<span class="sd">        genome: list of codons in genome</span>
<span class="sd">        grammar: BNF grammar used to create individual phenotype</span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>


<span class="sd">    Returns:</span>
<span class="sd">        phenotype: </span>
<span class="sd">        nodes: number of nodes in tree</span>
<span class="sd">        depth: max depth of tree used in mapping</span>
<span class="sd">        used_codons: number of codons consumed in mapping</span>
<span class="sd">        invalid: True if valid phenotype not constructed</span>
<span class="sd">        int for number of wraps used (0 for this implementation as wrapping not used)</span>
<span class="sd">        structure: list of production rules chosen in mapping, </span>
<span class="sd">            used to monitor population diversity</span>
<span class="sd">        consumed_codons: 2D list containing max codon used in each chromosome</span>
<span class="sd">    &quot;&quot;&quot;</span>  

<span class="c1">#     idx_genome = 0</span>
    <span class="n">phenotype</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">start_rule</span>
    <span class="n">next_NT</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
    <span class="n">n_starting_NTs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="n">term</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)])</span>
    <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="n">n_starting_NTs</span> <span class="c1">#it keeps the depth of each branch</span>
    <span class="n">idx_depth</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nodes</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">structure</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># track position in each chromosome</span>
    <span class="n">consumed_codons</span> <span class="o">=</span> <span class="p">[</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>

    <span class="k">while</span> <span class="n">next_NT</span><span class="p">:</span>
        <span class="n">NT_index</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">next_NT</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">grammar</span><span class="o">.</span><span class="n">n_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#there is a single PR for this non-terminal</span>
            <span class="n">index_production_chosen</span> <span class="o">=</span> <span class="mi">0</span>        
        <span class="k">else</span><span class="p">:</span> <span class="c1">#we consume one codon, and add the index to the structure</span>
            <span class="n">chr_idx</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span>
            <span class="c1"># if the end of a chromosome has been reached, stop mapping</span>
            <span class="k">if</span><span class="p">(</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">chr_idx</span><span class="p">]</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="n">chr_idx</span><span class="p">])):</span>
                <span class="k">break</span>

            <span class="n">index_production_chosen</span> <span class="o">=</span> <span class="n">genome</span><span class="p">[</span><span class="n">chr_idx</span><span class="p">][</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">chr_idx</span><span class="p">]]</span> <span class="o">%</span> <span class="n">grammar</span><span class="o">.</span><span class="n">n_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">]</span>
            <span class="n">consumed_codons</span><span class="p">[</span><span class="n">chr_idx</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">structure</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">index_production_chosen</span><span class="p">)</span>

        <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">next_NT</span><span class="p">,</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
            <span class="k">break</span>
        <span class="k">if</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> <span class="c1">#arity 0 (T)</span>
            <span class="n">idx_depth</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">nodes</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span> <span class="c1">#arity 1 (PR with one NT)</span>
            <span class="k">pass</span>        
        <span class="k">else</span><span class="p">:</span> <span class="c1">#it is a PR with more than one NT</span>
            <span class="n">arity</span> <span class="o">=</span> <span class="n">grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">NT_index</span><span class="p">][</span><span class="n">index_production_chosen</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">idx_depth</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">list_depth</span> <span class="o">=</span> <span class="n">list_depth</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_depth</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="p">],]</span><span class="o">*</span><span class="n">arity</span> <span class="o">+</span> <span class="n">list_depth</span><span class="p">[</span><span class="n">idx_depth</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>

        <span class="n">next_</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;(\w+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">next_</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="n">next_</span><span class="o">.</span><span class="n">group</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">next_NT</span> <span class="o">=</span> <span class="kc">None</span>


    <span class="k">if</span> <span class="n">next_NT</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">consumed_codons</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">invalid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">used_codons</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">consumed_codons</span><span class="p">)</span>

    <span class="n">depth</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">list_depth</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">phenotype</span><span class="p">,</span> <span class="n">nodes</span><span class="p">,</span> <span class="n">depth</span><span class="p">,</span> <span class="n">used_codons</span><span class="p">,</span> <span class="n">invalid</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">structure</span><span class="p">,</span> <span class="n">consumed_codons</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mcge_crossover_onepoint" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mcge_crossover_onepoint</span><span class="p">(</span><span class="n">parent0</span><span class="p">,</span> <span class="n">parent1</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">genome_representation</span><span class="o">=</span><span class="s1">&#39;list&#39;</span><span class="p">,</span> <span class="n">max_genome_length</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>One point crossover for standard LEAP GE genome individuals. Crossovers occur
    between a matching pair of chromosomes in each genome</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>parent0</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>individual to cross</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>parent1</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>second individual to cross</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genome_representation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>either list or numpy</p>
              </div>
            </td>
            <td>
                  <code>&#39;list&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum allowed length of genome when not None</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>new_ind0</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new individual resulting from crossover</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>new_ind1</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>second new individual resulting from crossover</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1711</span>
<span class="normal">1712</span>
<span class="normal">1713</span>
<span class="normal">1714</span>
<span class="normal">1715</span>
<span class="normal">1716</span>
<span class="normal">1717</span>
<span class="normal">1718</span>
<span class="normal">1719</span>
<span class="normal">1720</span>
<span class="normal">1721</span>
<span class="normal">1722</span>
<span class="normal">1723</span>
<span class="normal">1724</span>
<span class="normal">1725</span>
<span class="normal">1726</span>
<span class="normal">1727</span>
<span class="normal">1728</span>
<span class="normal">1729</span>
<span class="normal">1730</span>
<span class="normal">1731</span>
<span class="normal">1732</span>
<span class="normal">1733</span>
<span class="normal">1734</span>
<span class="normal">1735</span>
<span class="normal">1736</span>
<span class="normal">1737</span>
<span class="normal">1738</span>
<span class="normal">1739</span>
<span class="normal">1740</span>
<span class="normal">1741</span>
<span class="normal">1742</span>
<span class="normal">1743</span>
<span class="normal">1744</span>
<span class="normal">1745</span>
<span class="normal">1746</span>
<span class="normal">1747</span>
<span class="normal">1748</span>
<span class="normal">1749</span>
<span class="normal">1750</span>
<span class="normal">1751</span>
<span class="normal">1752</span>
<span class="normal">1753</span>
<span class="normal">1754</span>
<span class="normal">1755</span>
<span class="normal">1756</span>
<span class="normal">1757</span>
<span class="normal">1758</span>
<span class="normal">1759</span>
<span class="normal">1760</span>
<span class="normal">1761</span>
<span class="normal">1762</span>
<span class="normal">1763</span>
<span class="normal">1764</span>
<span class="normal">1765</span>
<span class="normal">1766</span>
<span class="normal">1767</span>
<span class="normal">1768</span>
<span class="normal">1769</span>
<span class="normal">1770</span>
<span class="normal">1771</span>
<span class="normal">1772</span>
<span class="normal">1773</span>
<span class="normal">1774</span>
<span class="normal">1775</span>
<span class="normal">1776</span>
<span class="normal">1777</span>
<span class="normal">1778</span>
<span class="normal">1779</span>
<span class="normal">1780</span>
<span class="normal">1781</span>
<span class="normal">1782</span>
<span class="normal">1783</span>
<span class="normal">1784</span>
<span class="normal">1785</span>
<span class="normal">1786</span>
<span class="normal">1787</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mcge_crossover_onepoint</span><span class="p">(</span><span class="n">parent0</span><span class="p">:</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> <span class="n">parent1</span><span class="p">:</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> 
                       <span class="n">bnf_grammar</span><span class="p">:</span> <span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
                    <span class="n">genome_representation</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s1">&#39;list&#39;</span><span class="p">,</span> 
                    <span class="n">max_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span><span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;One point crossover for standard LEAP GE genome individuals. Crossovers occur</span>
<span class="sd">        between a matching pair of chromosomes in each genome</span>

<span class="sd">    Args:</span>
<span class="sd">        parent0: individual to cross</span>
<span class="sd">        parent1: second individual to cross</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        genome_representation: either list or numpy</span>
<span class="sd">        max_genome_length: maximum allowed length of genome when not None</span>

<span class="sd">    Returns:</span>
<span class="sd">        new_ind0: new individual resulting from crossover</span>
<span class="sd">        new_ind1: second new individual resulting from crossover</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="n">idx_chr</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">parent0</span><span class="o">.</span><span class="n">invalid</span> <span class="ow">or</span> <span class="n">parent0</span><span class="o">.</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span> 
        <span class="n">possible_crossover_codons0</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">possible_crossover_codons0</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">]),</span> <span class="n">parent0</span><span class="o">.</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">])</span> <span class="c1">#in case of wrapping, used_codons can be greater than genome&#39;s length</span>
    <span class="k">if</span> <span class="n">parent1</span><span class="o">.</span><span class="n">invalid</span> <span class="ow">or</span> <span class="n">parent1</span><span class="o">.</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">possible_crossover_codons1</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">possible_crossover_codons1</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">]),</span> <span class="n">parent1</span><span class="o">.</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">])</span>

    <span class="c1">#Set points for crossover within the effective part of the genomes</span>
    <span class="n">point0</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">possible_crossover_codons0</span><span class="p">)</span>
    <span class="n">point1</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">possible_crossover_codons1</span><span class="p">)</span>


    <span class="k">if</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;numpy&#39;</span><span class="p">:</span>
        <span class="c1">#TODO Update this code to work with MCGE</span>
        <span class="c1">#TODO This operations is not working in case of wrapping</span>
        <span class="n">len0</span> <span class="o">=</span> <span class="n">point0</span> <span class="o">+</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">-</span> <span class="n">point1</span><span class="p">)</span>
        <span class="n">len1</span> <span class="o">=</span> <span class="n">point1</span> <span class="o">+</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">-</span> <span class="n">point0</span><span class="p">)</span>
        <span class="n">new_genome0</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">([</span><span class="n">len0</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">new_genome1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">([</span><span class="n">len1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">int</span><span class="p">)</span>

        <span class="c1">#Operate crossover</span>
        <span class="n">new_genome0</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point0</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point0</span><span class="p">]</span>
        <span class="n">new_genome0</span><span class="p">[</span><span class="n">point0</span><span class="p">:]</span> <span class="o">=</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">point1</span><span class="p">:]</span>
        <span class="n">new_genome1</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point1</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">point1</span><span class="p">]</span>
        <span class="n">new_genome1</span><span class="p">[</span><span class="n">point1</span><span class="p">:]</span> <span class="o">=</span> <span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">point0</span><span class="p">:]</span>

    <span class="k">elif</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
        <span class="c1">#Operate crossover - copy parents and then cross over designated chromosome</span>
        <span class="n">new_genome0</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        <span class="n">new_genome1</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        <span class="n">new_genome0</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="n">point0</span><span class="p">]</span> <span class="o">+</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">][</span><span class="n">point1</span><span class="p">:]</span>
        <span class="n">new_genome1</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent1</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">][</span><span class="mi">0</span><span class="p">:</span><span class="n">point1</span><span class="p">]</span> <span class="o">+</span> <span class="n">parent0</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">][</span><span class="n">point0</span><span class="p">:]</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown genome representation&quot;</span><span class="p">)</span>


    <span class="n">new_ind0</span> <span class="o">=</span> <span class="n">reMap_mcge</span><span class="p">(</span><span class="n">parent0</span><span class="p">,</span> <span class="n">new_genome0</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>
    <span class="n">new_ind1</span> <span class="o">=</span> <span class="n">reMap_mcge</span><span class="p">(</span><span class="n">parent1</span><span class="p">,</span> <span class="n">new_genome1</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">new_ind0</span><span class="o">.</span><span class="n">depth</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
        <span class="n">new_ind0</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">if</span> <span class="n">new_ind1</span><span class="o">.</span><span class="n">depth</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
        <span class="n">new_ind1</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">max_genome_length</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_ind0</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_genome_length</span><span class="p">:</span>
            <span class="n">new_ind0</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_ind1</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_genome_length</span><span class="p">:</span>
            <span class="n">new_ind1</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">return</span> <span class="n">new_ind0</span><span class="p">,</span> <span class="n">new_ind1</span> 
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mcge_mutation_int_flip_per_codon" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mcge_mutation_int_flip_per_codon</span><span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">mut_probability</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">max_genome_length</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Mutation operator for multi-chromosome GE genome. Each codon within the effective used codons
of each chromosome are checked</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>individual to mutate</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>mut_probability</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>chance for each codon to mutate</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum value for a codon</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum allowed length of genome when not None</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>new_ind</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new individual incorporating any mutations</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1893</span>
<span class="normal">1894</span>
<span class="normal">1895</span>
<span class="normal">1896</span>
<span class="normal">1897</span>
<span class="normal">1898</span>
<span class="normal">1899</span>
<span class="normal">1900</span>
<span class="normal">1901</span>
<span class="normal">1902</span>
<span class="normal">1903</span>
<span class="normal">1904</span>
<span class="normal">1905</span>
<span class="normal">1906</span>
<span class="normal">1907</span>
<span class="normal">1908</span>
<span class="normal">1909</span>
<span class="normal">1910</span>
<span class="normal">1911</span>
<span class="normal">1912</span>
<span class="normal">1913</span>
<span class="normal">1914</span>
<span class="normal">1915</span>
<span class="normal">1916</span>
<span class="normal">1917</span>
<span class="normal">1918</span>
<span class="normal">1919</span>
<span class="normal">1920</span>
<span class="normal">1921</span>
<span class="normal">1922</span>
<span class="normal">1923</span>
<span class="normal">1924</span>
<span class="normal">1925</span>
<span class="normal">1926</span>
<span class="normal">1927</span>
<span class="normal">1928</span>
<span class="normal">1929</span>
<span class="normal">1930</span>
<span class="normal">1931</span>
<span class="normal">1932</span>
<span class="normal">1933</span>
<span class="normal">1934</span>
<span class="normal">1935</span>
<span class="normal">1936</span>
<span class="normal">1937</span>
<span class="normal">1938</span>
<span class="normal">1939</span>
<span class="normal">1940</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mcge_mutation_int_flip_per_codon</span><span class="p">(</span><span class="n">ind</span><span class="p">:</span><span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> <span class="n">mut_probability</span><span class="p">:</span><span class="nb">float</span><span class="p">,</span> 
                                <span class="n">codon_size</span><span class="p">:</span><span class="nb">int</span> <span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
                                <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
                                <span class="n">max_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Mutation operator for multi-chromosome GE genome. Each codon within the effective used codons</span>
<span class="sd">    of each chromosome are checked</span>

<span class="sd">    Args:</span>
<span class="sd">        ind: individual to mutate</span>
<span class="sd">        mut_probability: chance for each codon to mutate</span>
<span class="sd">        codon_size: maximum value for a codon</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        max_genome_length: maximum allowed length of genome when not None</span>

<span class="sd">    Returns:</span>
<span class="sd">        new_ind: new individual incorporating any mutations</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="c1"># Operation mutation within the effective part of the genome</span>
    <span class="n">mutated_</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">idx_chr</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">ind</span><span class="o">.</span><span class="n">invalid</span><span class="p">:</span> <span class="c1">#used_codons = 0</span>
            <span class="n">possible_mutation_codons</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">possible_mutation_codons</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">]),</span> <span class="n">ind</span><span class="o">.</span><span class="n">consumed_codons</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">])</span> 

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">possible_mutation_codons</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">mut_probability</span><span class="p">:</span>
                <span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">idx_chr</span><span class="p">][</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">)</span>
                <span class="n">mutated_</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">mutated_</span><span class="p">:</span>
        <span class="n">new_ind</span> <span class="o">=</span> <span class="n">reMap_mcge</span><span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">new_ind</span> <span class="o">=</span> <span class="n">ind</span>

    <span class="k">if</span> <span class="n">new_ind</span><span class="o">.</span><span class="n">depth</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
        <span class="n">new_ind</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">max_genome_length</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_genome_length</span><span class="p">:</span>
            <span class="n">new_ind</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">mutated_</span><span class="p">:</span>
        <span class="k">del</span> <span class="n">new_ind</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span>
    <span class="k">return</span> <span class="n">new_ind</span><span class="p">,</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mcge_random_initializaion" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mcge_random_initializaion</span><span class="p">(</span><span class="n">ind_class</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">min_init_genome_length</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">genome_representation</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Randomly generated linear genome compatible with MCGE chromosome mapping. 
Once chromosome created for each Non-terminal</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind_class</code></td>
            <td>
                  <code><span title="deap.creator.MetaCreator">MetaCreator</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Deap creator</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>pop_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of individuals to create</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>min_init_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>minimum number of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum number of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth for tree in mapping genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum to be stored in codon</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genome_representation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>either list or numpy</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>population</code></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of individuals</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mcge_random_initializaion</span><span class="p">(</span><span class="n">ind_class</span><span class="p">:</span> <span class="s1">&#39;deap.creator.MetaCreator&#39;</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
                          <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> 
                          <span class="n">min_init_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">max_init_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
                          <span class="n">genome_representation</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Randomly generated linear genome compatible with MCGE chromosome mapping. </span>
<span class="sd">    Once chromosome created for each Non-terminal</span>

<span class="sd">    Args:</span>
<span class="sd">        ind_class: Deap creator</span>
<span class="sd">        pop_size: number of individuals to create</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        min_init_genome_length: minimum number of codons in genome</span>
<span class="sd">        max_init_genome_length: maximum number of codons in genome</span>
<span class="sd">        max_init_depth: maximum depth for tree in mapping genome</span>
<span class="sd">        codon_size: maximum to be stored in codon</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        genome_representation: either list or numpy</span>


<span class="sd">    Returns:</span>
<span class="sd">        population: list of individuals</span>
<span class="sd">    &quot;&quot;&quot;</span> 
    <span class="n">population</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">pop_size</span><span class="p">):</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">):</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">init_genome_length</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">min_init_genome_length</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">init_genome_length</span><span class="p">):</span>
                <span class="n">chrom</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">))</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span>

        <span class="n">ind</span> <span class="o">=</span> <span class="n">ind_class</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>
        <span class="n">population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">elif</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;numpy&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">population</span><span class="p">:</span>
            <span class="n">ind</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unkonwn genome representation&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mcge_sensible_initialisation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mcge_sensible_initialisation</span><span class="p">(</span><span class="n">ind_class</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">min_init_genome_length</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">genome_representation</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Sensible initialization for MCGE genome. Half the generated individuals
are up to the depth of the max_init_depth and half are limited to that depth but may be
smaller</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind_class</code></td>
            <td>
                  <code><span title="deap.creator.MetaCreator">MetaCreator</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Deap creator</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>pop_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of individuals to create</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>min_init_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>minimum number of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum number of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth for tree in mapping genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum to be stored in codon</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genome_representation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>either list or numpy</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>population</code></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of individuals</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1394</span>
<span class="normal">1395</span>
<span class="normal">1396</span>
<span class="normal">1397</span>
<span class="normal">1398</span>
<span class="normal">1399</span>
<span class="normal">1400</span>
<span class="normal">1401</span>
<span class="normal">1402</span>
<span class="normal">1403</span>
<span class="normal">1404</span>
<span class="normal">1405</span>
<span class="normal">1406</span>
<span class="normal">1407</span>
<span class="normal">1408</span>
<span class="normal">1409</span>
<span class="normal">1410</span>
<span class="normal">1411</span>
<span class="normal">1412</span>
<span class="normal">1413</span>
<span class="normal">1414</span>
<span class="normal">1415</span>
<span class="normal">1416</span>
<span class="normal">1417</span>
<span class="normal">1418</span>
<span class="normal">1419</span>
<span class="normal">1420</span>
<span class="normal">1421</span>
<span class="normal">1422</span>
<span class="normal">1423</span>
<span class="normal">1424</span>
<span class="normal">1425</span>
<span class="normal">1426</span>
<span class="normal">1427</span>
<span class="normal">1428</span>
<span class="normal">1429</span>
<span class="normal">1430</span>
<span class="normal">1431</span>
<span class="normal">1432</span>
<span class="normal">1433</span>
<span class="normal">1434</span>
<span class="normal">1435</span>
<span class="normal">1436</span>
<span class="normal">1437</span>
<span class="normal">1438</span>
<span class="normal">1439</span>
<span class="normal">1440</span>
<span class="normal">1441</span>
<span class="normal">1442</span>
<span class="normal">1443</span>
<span class="normal">1444</span>
<span class="normal">1445</span>
<span class="normal">1446</span>
<span class="normal">1447</span>
<span class="normal">1448</span>
<span class="normal">1449</span>
<span class="normal">1450</span>
<span class="normal">1451</span>
<span class="normal">1452</span>
<span class="normal">1453</span>
<span class="normal">1454</span>
<span class="normal">1455</span>
<span class="normal">1456</span>
<span class="normal">1457</span>
<span class="normal">1458</span>
<span class="normal">1459</span>
<span class="normal">1460</span>
<span class="normal">1461</span>
<span class="normal">1462</span>
<span class="normal">1463</span>
<span class="normal">1464</span>
<span class="normal">1465</span>
<span class="normal">1466</span>
<span class="normal">1467</span>
<span class="normal">1468</span>
<span class="normal">1469</span>
<span class="normal">1470</span>
<span class="normal">1471</span>
<span class="normal">1472</span>
<span class="normal">1473</span>
<span class="normal">1474</span>
<span class="normal">1475</span>
<span class="normal">1476</span>
<span class="normal">1477</span>
<span class="normal">1478</span>
<span class="normal">1479</span>
<span class="normal">1480</span>
<span class="normal">1481</span>
<span class="normal">1482</span>
<span class="normal">1483</span>
<span class="normal">1484</span>
<span class="normal">1485</span>
<span class="normal">1486</span>
<span class="normal">1487</span>
<span class="normal">1488</span>
<span class="normal">1489</span>
<span class="normal">1490</span>
<span class="normal">1491</span>
<span class="normal">1492</span>
<span class="normal">1493</span>
<span class="normal">1494</span>
<span class="normal">1495</span>
<span class="normal">1496</span>
<span class="normal">1497</span>
<span class="normal">1498</span>
<span class="normal">1499</span>
<span class="normal">1500</span>
<span class="normal">1501</span>
<span class="normal">1502</span>
<span class="normal">1503</span>
<span class="normal">1504</span>
<span class="normal">1505</span>
<span class="normal">1506</span>
<span class="normal">1507</span>
<span class="normal">1508</span>
<span class="normal">1509</span>
<span class="normal">1510</span>
<span class="normal">1511</span>
<span class="normal">1512</span>
<span class="normal">1513</span>
<span class="normal">1514</span>
<span class="normal">1515</span>
<span class="normal">1516</span>
<span class="normal">1517</span>
<span class="normal">1518</span>
<span class="normal">1519</span>
<span class="normal">1520</span>
<span class="normal">1521</span>
<span class="normal">1522</span>
<span class="normal">1523</span>
<span class="normal">1524</span>
<span class="normal">1525</span>
<span class="normal">1526</span>
<span class="normal">1527</span>
<span class="normal">1528</span>
<span class="normal">1529</span>
<span class="normal">1530</span>
<span class="normal">1531</span>
<span class="normal">1532</span>
<span class="normal">1533</span>
<span class="normal">1534</span>
<span class="normal">1535</span>
<span class="normal">1536</span>
<span class="normal">1537</span>
<span class="normal">1538</span>
<span class="normal">1539</span>
<span class="normal">1540</span>
<span class="normal">1541</span>
<span class="normal">1542</span>
<span class="normal">1543</span>
<span class="normal">1544</span>
<span class="normal">1545</span>
<span class="normal">1546</span>
<span class="normal">1547</span>
<span class="normal">1548</span>
<span class="normal">1549</span>
<span class="normal">1550</span>
<span class="normal">1551</span>
<span class="normal">1552</span>
<span class="normal">1553</span>
<span class="normal">1554</span>
<span class="normal">1555</span>
<span class="normal">1556</span>
<span class="normal">1557</span>
<span class="normal">1558</span>
<span class="normal">1559</span>
<span class="normal">1560</span>
<span class="normal">1561</span>
<span class="normal">1562</span>
<span class="normal">1563</span>
<span class="normal">1564</span>
<span class="normal">1565</span>
<span class="normal">1566</span>
<span class="normal">1567</span>
<span class="normal">1568</span>
<span class="normal">1569</span>
<span class="normal">1570</span>
<span class="normal">1571</span>
<span class="normal">1572</span>
<span class="normal">1573</span>
<span class="normal">1574</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mcge_sensible_initialisation</span><span class="p">(</span><span class="n">ind_class</span><span class="p">:</span> <span class="s1">&#39;deap.creator.MetaCreator&#39;</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
                          <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> 
                          <span class="n">min_init_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">max_init_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
                          <span class="n">genome_representation</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Sensible initialization for MCGE genome. Half the generated individuals</span>
<span class="sd">    are up to the depth of the max_init_depth and half are limited to that depth but may be</span>
<span class="sd">    smaller</span>

<span class="sd">    Args:</span>
<span class="sd">        ind_class: Deap creator</span>
<span class="sd">        pop_size: number of individuals to create</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        min_init_genome_length: minimum number of codons in genome</span>
<span class="sd">        max_init_genome_length: maximum number of codons in genome</span>
<span class="sd">        max_init_depth: maximum depth for tree in mapping genome</span>
<span class="sd">        codon_size: maximum to be stored in codon</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        genome_representation: either list or numpy</span>


<span class="sd">    Returns:</span>
<span class="sd">        population: list of individuals</span>
<span class="sd">    &quot;&quot;&quot;</span> 
    <span class="c1">#Calculate the number of individuals to be generated with each method</span>
    <span class="n">is_odd</span> <span class="o">=</span> <span class="n">pop_size</span> <span class="o">%</span> <span class="mi">2</span>
    <span class="n">n_grow</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pop_size</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>

    <span class="n">n_sets_grow</span> <span class="o">=</span> <span class="n">max_init_depth</span> <span class="o">-</span> <span class="n">min_init_depth</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">set_size</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_grow</span><span class="o">/</span><span class="n">n_sets_grow</span><span class="p">)</span>
    <span class="n">remaining</span> <span class="o">=</span> <span class="n">n_grow</span> <span class="o">%</span> <span class="n">n_sets_grow</span>

    <span class="n">n_full</span> <span class="o">=</span> <span class="n">n_grow</span> <span class="o">+</span> <span class="n">is_odd</span> <span class="o">+</span> <span class="n">remaining</span> <span class="c1">#if pop_size is odd, generate an extra ind with &quot;full&quot;</span>

    <span class="c1">#TODO check if it is possible to generate inds with max_init_depth</span>
    <span class="c1"># this has not been done as I discovered before debugging my problem with GENN grammar</span>

    <span class="n">population</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1">#Generate inds using &quot;Grow&quot;</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_sets_grow</span><span class="p">):</span>
        <span class="n">max_init_depth_</span> <span class="o">=</span> <span class="n">min_init_depth</span> <span class="o">+</span> <span class="n">i</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">set_size</span><span class="p">):</span>
            <span class="n">remainders</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the choices</span>
            <span class="n">possible_choices</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the respective possible choices</span>
            <span class="n">rules_used</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#registers which rule used for setting the frame</span>

            <span class="n">phenotype</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">start_rule</span>
            <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span> <span class="c1">#</span>
            <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="c1">#it keeps the depth of each branch</span>
            <span class="n">idx_branch</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1">#index of the current branch being grown</span>
            <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">idx_NT</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">total_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]]</span>
                <span class="n">actual_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]</span> <span class="k">if</span> <span class="n">PR</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_init_depth_</span><span class="p">]</span>
                <span class="n">Ch</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">actual_options</span><span class="p">)</span>
                <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
                <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
                    <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                    <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>
                    <span class="n">rules_used</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx_NT</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                        <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>
                        <span class="n">rules_used</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx_NT</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">idx_branch</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">depths</span> <span class="o">=</span> <span class="n">depths</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
                <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;terminal&#39;</span><span class="p">:</span>
                    <span class="n">idx_branch</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span>

            <span class="c1">#Generate the genome</span>
            <span class="c1">#remainders contain the choice for the rule</span>
            <span class="c1">#possible_choices simply tells you how many choices for that codon -- needed to generate actual codon value</span>
            <span class="c1"># self.rule_conversion.append(idx)</span>
            <span class="c1"># self.nt_rule_size += 1</span>

            <span class="c1"># set up 2-D list </span>
            <span class="n">genome</span> <span class="o">=</span> <span class="p">[</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>
            <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span> <span class="ow">or</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">remainders</span><span class="p">)):</span>
                    <span class="n">codon</span> <span class="o">=</span> <span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">int</span><span class="p">(</span><span class="mf">1e10</span><span class="p">))</span> <span class="o">%</span> <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(((</span><span class="n">codon_size</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span> <span class="o">*</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">k</span><span class="p">])</span> <span class="o">+</span> <span class="n">remainders</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
                    <span class="n">genome</span><span class="p">[</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">[</span><span class="n">rules_used</span><span class="p">[</span><span class="n">k</span><span class="p">]]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codon</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>

            <span class="c1">#Include a tail with 50% of the genome&#39;s size</span>
            <span class="c1"># calculate based on number of frames and extend by complete frames</span>
            <span class="c1"># add tail to each chromosome</span>
            <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">:</span>
                <span class="n">size_tail</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.5</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">chrom</span><span class="p">)),</span> <span class="mi">10</span><span class="p">)</span> <span class="c1">#Tail must have at least one codon. Otherwise, in the lazy approach, when we have the last PR with just a single option, the mapping procces will not terminate.</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">size_tail</span><span class="p">):</span>
                    <span class="n">chrom</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">codon_size</span><span class="p">))</span>

            <span class="c1">#Initialise the individual and include in the population</span>
            <span class="n">ind</span> <span class="o">=</span> <span class="n">ind_class</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_init_depth_</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>

            <span class="c1">#Check if the individual was mapped correctly</span>
            <span class="k">if</span> <span class="n">remainders</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">structure</span> <span class="ow">or</span> <span class="n">phenotype</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">phenotype</span> <span class="ow">or</span> <span class="nb">max</span><span class="p">(</span><span class="n">depths</span><span class="p">)</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">depth</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;error in the mapping&#39;</span><span class="p">)</span>
            <span class="n">population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>    

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_full</span><span class="p">):</span>
        <span class="n">remainders</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the choices</span>
        <span class="n">possible_choices</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the respective possible choices</span>
        <span class="n">rules_used</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#registers which rule used for setting the frame</span>

        <span class="n">phenotype</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">start_rule</span>
        <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span> <span class="c1">#</span>
        <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="c1">#it keeps the depth of each branch</span>
        <span class="n">idx_branch</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1">#index of the current branch being grown</span>

        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">idx_NT</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">total_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]]</span>
            <span class="n">actual_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]</span> <span class="k">if</span> <span class="n">PR</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_init_depth</span><span class="p">]</span>
            <span class="n">recursive_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">actual_options</span> <span class="k">if</span> <span class="n">PR</span><span class="p">[</span><span class="mi">4</span><span class="p">]]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">recursive_options</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">Ch</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">recursive_options</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">Ch</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">actual_options</span><span class="p">)</span>
            <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
                <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>
                <span class="n">rules_used</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx_NT</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                    <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>
                    <span class="n">rules_used</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx_NT</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">idx_branch</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">depths</span> <span class="o">=</span> <span class="n">depths</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;terminal&#39;</span><span class="p">:</span>
                <span class="n">idx_branch</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span>

        <span class="c1">#Generate the genome</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="p">[</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">nt_rule_size</span><span class="p">)]</span>
        <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span> <span class="ow">or</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">remainders</span><span class="p">)):</span>
                <span class="n">codon</span> <span class="o">=</span> <span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">int</span><span class="p">(</span><span class="mf">1e10</span><span class="p">))</span> <span class="o">%</span> <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(((</span><span class="n">codon_size</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">j</span><span class="p">]))</span> <span class="o">*</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">j</span><span class="p">])</span> <span class="o">+</span> <span class="n">remainders</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                <span class="n">genome</span><span class="p">[</span><span class="n">bnf_grammar</span><span class="o">.</span><span class="n">rule_conversion</span><span class="p">[</span><span class="n">rules_used</span><span class="p">[</span><span class="n">j</span><span class="p">]]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codon</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>

        <span class="c1">#Include a tail with 50% of the genome&#39;s size</span>
        <span class="k">for</span> <span class="n">chrom</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">:</span>
            <span class="n">size_tail</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.5</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">chrom</span><span class="p">)),</span> <span class="mi">10</span><span class="p">)</span> <span class="c1">#Tail must have at least one codon. Otherwise, in the lazy approach, when we have the last PR with just a single option, the mapping procces will not terminate.</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">size_tail</span><span class="p">):</span>
                <span class="n">chrom</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">codon_size</span><span class="p">))</span>
        <span class="c1">#Initialise the individual and include in the population</span>
        <span class="n">ind</span> <span class="o">=</span> <span class="n">ind_class</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>

        <span class="c1">#Check if the individual was mapped correctly</span>
        <span class="k">if</span> <span class="n">remainders</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">structure</span> <span class="ow">or</span> <span class="n">phenotype</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">phenotype</span> <span class="ow">or</span> <span class="nb">max</span><span class="p">(</span><span class="n">depths</span><span class="p">)</span> <span class="o">!=</span> <span class="n">ind</span><span class="o">.</span><span class="n">depth</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;error in the mapping&#39;</span><span class="p">)</span>

        <span class="n">population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>    

    <span class="k">if</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">elif</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;numpy&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">population</span><span class="p">:</span>
            <span class="n">ind</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unkonwn genome representation&quot;</span><span class="p">)</span>  
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.mutation_int_flip_per_codon" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mutation_int_flip_per_codon</span><span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">mut_probability</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">max_genome_length</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Mutation operator for standard linear GE genome. Each codon within the effective used codons
of the genome are checked</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>individual to mutate</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>mut_probability</code></td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>chance for each codon to mutate</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum value for a codon</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum allowed length of genome when not None</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>new_ind</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new individual incorporating any mutations</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1791</span>
<span class="normal">1792</span>
<span class="normal">1793</span>
<span class="normal">1794</span>
<span class="normal">1795</span>
<span class="normal">1796</span>
<span class="normal">1797</span>
<span class="normal">1798</span>
<span class="normal">1799</span>
<span class="normal">1800</span>
<span class="normal">1801</span>
<span class="normal">1802</span>
<span class="normal">1803</span>
<span class="normal">1804</span>
<span class="normal">1805</span>
<span class="normal">1806</span>
<span class="normal">1807</span>
<span class="normal">1808</span>
<span class="normal">1809</span>
<span class="normal">1810</span>
<span class="normal">1811</span>
<span class="normal">1812</span>
<span class="normal">1813</span>
<span class="normal">1814</span>
<span class="normal">1815</span>
<span class="normal">1816</span>
<span class="normal">1817</span>
<span class="normal">1818</span>
<span class="normal">1819</span>
<span class="normal">1820</span>
<span class="normal">1821</span>
<span class="normal">1822</span>
<span class="normal">1823</span>
<span class="normal">1824</span>
<span class="normal">1825</span>
<span class="normal">1826</span>
<span class="normal">1827</span>
<span class="normal">1828</span>
<span class="normal">1829</span>
<span class="normal">1830</span>
<span class="normal">1831</span>
<span class="normal">1832</span>
<span class="normal">1833</span>
<span class="normal">1834</span>
<span class="normal">1835</span>
<span class="normal">1836</span>
<span class="normal">1837</span>
<span class="normal">1838</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">mutation_int_flip_per_codon</span><span class="p">(</span><span class="n">ind</span><span class="p">:</span><span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> <span class="n">mut_probability</span><span class="p">:</span><span class="nb">float</span><span class="p">,</span> 
                                <span class="n">codon_size</span><span class="p">:</span><span class="nb">int</span> <span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
                                <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> 
                                <span class="n">max_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Mutation operator for standard linear GE genome. Each codon within the effective used codons</span>
<span class="sd">    of the genome are checked</span>

<span class="sd">    Args:</span>
<span class="sd">        ind: individual to mutate</span>
<span class="sd">        mut_probability: chance for each codon to mutate</span>
<span class="sd">        codon_size: maximum value for a codon</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        max_depth: maximum depth allowed for a tree during mapping</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        max_genome_length: maximum allowed length of genome when not None</span>

<span class="sd">    Returns:</span>
<span class="sd">        new_ind: new individual incorporating any mutations</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="c1"># Operation mutation within the effective part of the genome</span>
    <span class="k">if</span> <span class="n">ind</span><span class="o">.</span><span class="n">invalid</span><span class="p">:</span> <span class="c1">#used_codons = 0</span>
        <span class="n">possible_mutation_codons</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">possible_mutation_codons</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">),</span> <span class="n">ind</span><span class="o">.</span><span class="n">used_codons</span><span class="p">)</span> <span class="c1">#in case of wrapping, used_codons can be greater than genome&#39;s length</span>

    <span class="n">mutated_</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">possible_mutation_codons</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span> <span class="o">&lt;</span> <span class="n">mut_probability</span><span class="p">:</span>
            <span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">)</span>
            <span class="n">mutated_</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">mutated_</span><span class="p">:</span>
        <span class="n">new_ind</span> <span class="o">=</span> <span class="n">reMap</span><span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">new_ind</span> <span class="o">=</span> <span class="n">ind</span>

    <span class="k">if</span> <span class="n">new_ind</span><span class="o">.</span><span class="n">depth</span> <span class="o">&gt;</span> <span class="n">max_depth</span><span class="p">:</span>
        <span class="n">new_ind</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">max_genome_length</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">max_genome_length</span><span class="p">:</span>
            <span class="n">new_ind</span><span class="o">.</span><span class="n">invalid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="n">mutated_</span><span class="p">:</span>
        <span class="k">del</span> <span class="n">new_ind</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span>
    <span class="k">return</span> <span class="n">new_ind</span><span class="p">,</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.random_initialisation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">random_initialisation</span><span class="p">(</span><span class="n">ind_class</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">min_init_genome_length</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">genome_representation</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Randomly generated linear genome compatible with standard chromosome mapping</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind_class</code></td>
            <td>
                  <code><span title="deap.creator.MetaCreator">MetaCreator</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Deap creator</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>pop_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of individuals to create</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>min_init_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>minimum number of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_genome_length</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum number of codons in genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth for tree in mapping genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum to be stored in codon</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genome_representation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>either list or numpy</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>population</code></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of individuals</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">random_initialisation</span><span class="p">(</span><span class="n">ind_class</span><span class="p">:</span> <span class="s1">&#39;deap.creator.MetaCreator&#39;</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
                          <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> 
                          <span class="n">min_init_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">max_init_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
                          <span class="n">genome_representation</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Randomly generated linear genome compatible with standard chromosome mapping</span>

<span class="sd">    Args:</span>
<span class="sd">        ind_class: Deap creator</span>
<span class="sd">        pop_size: number of individuals to create</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        min_init_genome_length: minimum number of codons in genome</span>
<span class="sd">        max_init_genome_length: maximum number of codons in genome</span>
<span class="sd">        max_init_depth: maximum depth for tree in mapping genome</span>
<span class="sd">        codon_size: maximum to be stored in codon</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        genome_representation: either list or numpy</span>


<span class="sd">    Returns:</span>
<span class="sd">        population: list of individuals</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="n">population</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">pop_size</span><span class="p">):</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">init_genome_length</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">min_init_genome_length</span><span class="p">,</span> <span class="n">max_init_genome_length</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">init_genome_length</span><span class="p">):</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">))</span>
        <span class="n">ind</span> <span class="o">=</span> <span class="n">ind_class</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>
        <span class="n">population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">elif</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;numpy&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">population</span><span class="p">:</span>
            <span class="n">ind</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unkonwn genome representation&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.reMap" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">reMap</span><span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Maps standard GE genome in individual to produce new phenotype</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>individual to mutate</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_tree_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>new_ind</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new individual with updating phenotype and other mapping information</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1943</span>
<span class="normal">1944</span>
<span class="normal">1945</span>
<span class="normal">1946</span>
<span class="normal">1947</span>
<span class="normal">1948</span>
<span class="normal">1949</span>
<span class="normal">1950</span>
<span class="normal">1951</span>
<span class="normal">1952</span>
<span class="normal">1953</span>
<span class="normal">1954</span>
<span class="normal">1955</span>
<span class="normal">1956</span>
<span class="normal">1957</span>
<span class="normal">1958</span>
<span class="normal">1959</span>
<span class="normal">1960</span>
<span class="normal">1961</span>
<span class="normal">1962</span>
<span class="normal">1963</span>
<span class="normal">1964</span>
<span class="normal">1965</span>
<span class="normal">1966</span>
<span class="normal">1967</span>
<span class="normal">1968</span>
<span class="normal">1969</span>
<span class="normal">1970</span>
<span class="normal">1971</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">reMap</span><span class="p">(</span><span class="n">ind</span><span class="p">:</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> 
          <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
          <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Maps standard GE genome in individual to produce new phenotype</span>

<span class="sd">    Args:</span>
<span class="sd">        ind: individual to mutate</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        max_tree_depth: maximum depth allowed for a tree during mapping</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>

<span class="sd">    Returns:</span>
<span class="sd">        new_ind: new individual with updating phenotype and other mapping information</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="n">ind</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>
    <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
        <span class="n">ind</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_lazy</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
        <span class="n">ind</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_eager</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">)</span>
<span class="c1">#         ind.structure = mapper_eager_opt(genome, bnf_grammar, max_tree_depth)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>    

    <span class="k">return</span> <span class="n">ind</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.reMap_leap" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">reMap_leap</span><span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Maps LEAP GE genome in individual to produce new phenotype</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>individual to mutate</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_tree_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>new_ind</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new individual with updating phenotype and other mapping information</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1974</span>
<span class="normal">1975</span>
<span class="normal">1976</span>
<span class="normal">1977</span>
<span class="normal">1978</span>
<span class="normal">1979</span>
<span class="normal">1980</span>
<span class="normal">1981</span>
<span class="normal">1982</span>
<span class="normal">1983</span>
<span class="normal">1984</span>
<span class="normal">1985</span>
<span class="normal">1986</span>
<span class="normal">1987</span>
<span class="normal">1988</span>
<span class="normal">1989</span>
<span class="normal">1990</span>
<span class="normal">1991</span>
<span class="normal">1992</span>
<span class="normal">1993</span>
<span class="normal">1994</span>
<span class="normal">1995</span>
<span class="normal">1996</span>
<span class="normal">1997</span>
<span class="normal">1998</span>
<span class="normal">1999</span>
<span class="normal">2000</span>
<span class="normal">2001</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">reMap_leap</span><span class="p">(</span><span class="n">ind</span><span class="p">:</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> 
          <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
          <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Maps LEAP GE genome in individual to produce new phenotype</span>

<span class="sd">    Args:</span>
<span class="sd">        ind: individual to mutate</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        max_tree_depth: maximum depth allowed for a tree during mapping</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>

<span class="sd">    Returns:</span>
<span class="sd">        new_ind: new individual with updating phenotype and other mapping information</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="n">ind</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>
    <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
        <span class="n">ind</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_lazy_leap</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
        <span class="n">ind</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">structure</span> <span class="o">=</span> <span class="n">mapper_eager_leap</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ind</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.reMap_mcge" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">reMap_mcge</span><span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Maps multi-chromosome GE genome in individual to produce new phenotype</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind</code></td>
            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>individual to mutate</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_tree_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth allowed for a tree during mapping</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>new_ind</code></td>            <td>
                  <code><span title="deap.creator.Individual">Individual</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new individual with updating phenotype and other mapping information</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">2004</span>
<span class="normal">2005</span>
<span class="normal">2006</span>
<span class="normal">2007</span>
<span class="normal">2008</span>
<span class="normal">2009</span>
<span class="normal">2010</span>
<span class="normal">2011</span>
<span class="normal">2012</span>
<span class="normal">2013</span>
<span class="normal">2014</span>
<span class="normal">2015</span>
<span class="normal">2016</span>
<span class="normal">2017</span>
<span class="normal">2018</span>
<span class="normal">2019</span>
<span class="normal">2020</span>
<span class="normal">2021</span>
<span class="normal">2022</span>
<span class="normal">2023</span>
<span class="normal">2024</span>
<span class="normal">2025</span>
<span class="normal">2026</span>
<span class="normal">2027</span>
<span class="normal">2028</span>
<span class="normal">2029</span>
<span class="normal">2030</span>
<span class="normal">2031</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">reMap_mcge</span><span class="p">(</span><span class="n">ind</span><span class="p">:</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">,</span> <span class="n">genome</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> 
          <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
          <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s1">&#39;deap.creator.Individual&#39;</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Maps multi-chromosome GE genome in individual to produce new phenotype</span>

<span class="sd">    Args:</span>
<span class="sd">        ind: individual to mutate</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        max_tree_depth: maximum depth allowed for a tree during mapping</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>

<span class="sd">    Returns:</span>
<span class="sd">        new_ind: new individual with updating phenotype and other mapping information</span>
<span class="sd">    &quot;&quot;&quot;</span> 

    <span class="n">ind</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">genome</span>
    <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
        <span class="n">ind</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">structure</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">consumed_codons</span> <span class="o">=</span> <span class="n">mapper_lazy_mcge</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
        <span class="n">ind</span><span class="o">.</span><span class="n">phenotype</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">nodes</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">depth</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">used_codons</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">invalid</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">n_wraps</span><span class="p">,</span> \
        <span class="n">ind</span><span class="o">.</span><span class="n">structure</span><span class="p">,</span> <span class="n">ind</span><span class="o">.</span><span class="n">consumed_codons</span> <span class="o">=</span> <span class="n">mapper_eager_mcge</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_tree_depth</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ind</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.replace_nth" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">replace_nth</span><span class="p">(</span><span class="n">string</span><span class="p">,</span> <span class="n">substring</span><span class="p">,</span> <span class="n">new_substring</span><span class="p">,</span> <span class="n">nth</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>utility function to replace nth instance of a substring</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>string</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>original str</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>substring</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>substring to replace</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>new_substring</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>new substring to use</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>nth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>instance of occurrence of substring to replace</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns;
    updated string</p>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">2033</span>
<span class="normal">2034</span>
<span class="normal">2035</span>
<span class="normal">2036</span>
<span class="normal">2037</span>
<span class="normal">2038</span>
<span class="normal">2039</span>
<span class="normal">2040</span>
<span class="normal">2041</span>
<span class="normal">2042</span>
<span class="normal">2043</span>
<span class="normal">2044</span>
<span class="normal">2045</span>
<span class="normal">2046</span>
<span class="normal">2047</span>
<span class="normal">2048</span>
<span class="normal">2049</span>
<span class="normal">2050</span>
<span class="normal">2051</span>
<span class="normal">2052</span>
<span class="normal">2053</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">replace_nth</span><span class="p">(</span><span class="n">string</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">substring</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">new_substring</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">nth</span><span class="p">:</span><span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot; utility function to replace nth instance of a substring</span>

<span class="sd">    Args:</span>
<span class="sd">        string: original str</span>
<span class="sd">        substring: substring to replace</span>
<span class="sd">        new_substring: new substring to use</span>
<span class="sd">        nth: instance of occurrence of substring to replace</span>

<span class="sd">    Returns;</span>
<span class="sd">        updated string</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">find</span> <span class="o">=</span> <span class="n">string</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">substring</span><span class="p">)</span>
    <span class="n">i</span> <span class="o">=</span> <span class="n">find</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="k">while</span> <span class="n">find</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span> <span class="ow">and</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">nth</span><span class="p">:</span>
        <span class="n">find</span> <span class="o">=</span> <span class="n">string</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">substring</span><span class="p">,</span> <span class="n">find</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">nth</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">string</span><span class="p">[:</span><span class="n">find</span><span class="p">]</span> <span class="o">+</span> <span class="n">new_substring</span> <span class="o">+</span> <span class="n">string</span><span class="p">[</span><span class="n">find</span><span class="o">+</span><span class="nb">len</span><span class="p">(</span><span class="n">substring</span><span class="p">):]</span>
    <span class="k">return</span> <span class="n">string</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.selAthenaRoulette" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">selAthenaRoulette</span><span class="p">(</span><span class="n">individuals</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">fit_attr</span><span class="o">=</span><span class="s1">&#39;fitness&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Matches the original C++ ATHENA roulette selection method</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>individuals</code></td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of individuals to select from.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of individuals to select.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>fit_attr</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The attribute of individuals to use as selection criterion</p>
              </div>
            </td>
            <td>
                  <code>&#39;fitness&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list containing the k selected individuals</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">2079</span>
<span class="normal">2080</span>
<span class="normal">2081</span>
<span class="normal">2082</span>
<span class="normal">2083</span>
<span class="normal">2084</span>
<span class="normal">2085</span>
<span class="normal">2086</span>
<span class="normal">2087</span>
<span class="normal">2088</span>
<span class="normal">2089</span>
<span class="normal">2090</span>
<span class="normal">2091</span>
<span class="normal">2092</span>
<span class="normal">2093</span>
<span class="normal">2094</span>
<span class="normal">2095</span>
<span class="normal">2096</span>
<span class="normal">2097</span>
<span class="normal">2098</span>
<span class="normal">2099</span>
<span class="normal">2100</span>
<span class="normal">2101</span>
<span class="normal">2102</span>
<span class="normal">2103</span>
<span class="normal">2104</span>
<span class="normal">2105</span>
<span class="normal">2106</span>
<span class="normal">2107</span>
<span class="normal">2108</span>
<span class="normal">2109</span>
<span class="normal">2110</span>
<span class="normal">2111</span>
<span class="normal">2112</span>
<span class="normal">2113</span>
<span class="normal">2114</span>
<span class="normal">2115</span>
<span class="normal">2116</span>
<span class="normal">2117</span>
<span class="normal">2118</span>
<span class="normal">2119</span>
<span class="normal">2120</span>
<span class="normal">2121</span>
<span class="normal">2122</span>
<span class="normal">2123</span>
<span class="normal">2124</span>
<span class="normal">2125</span>
<span class="normal">2126</span>
<span class="normal">2127</span>
<span class="normal">2128</span>
<span class="normal">2129</span>
<span class="normal">2130</span>
<span class="normal">2131</span>
<span class="normal">2132</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">selAthenaRoulette</span><span class="p">(</span><span class="n">individuals</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">k</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">fit_attr</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s2">&quot;fitness&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Matches the original C++ ATHENA roulette selection method</span>

<span class="sd">    Args:</span>
<span class="sd">        individuals: A list of individuals to select from.</span>
<span class="sd">        k: The number of individuals to select.</span>
<span class="sd">        fit_attr: The attribute of individuals to use as selection criterion</span>

<span class="sd">    Returns: </span>
<span class="sd">        A list containing the k selected individuals</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># sort the population in ascending order</span>
    <span class="n">individuals</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">individuals</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="n">fit_attr</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">psum</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">individuals</span><span class="p">)</span>

    <span class="c1"># set the probabilities for each i ndividual to be selected</span>
    <span class="c1"># if all identical then make all probabilities equal</span>
    <span class="k">if</span> <span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">individuals</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
        <span class="n">psum</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">fitness</span><span class="o">/</span><span class="nb">float</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">individuals</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">maxfit</span> <span class="o">=</span> <span class="n">individuals</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">minfit</span> <span class="o">=</span> <span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="c1"># set first probability</span>
        <span class="n">psum</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="o">-</span><span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">maxfit</span> <span class="o">+</span> <span class="n">minfit</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">ind</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">individuals</span><span class="p">[</span><span class="mi">1</span><span class="p">:],</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">psum</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="o">-</span><span class="n">ind</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">minfit</span> <span class="o">+</span> <span class="n">maxfit</span> <span class="o">+</span> <span class="n">psum</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
            <span class="n">psum</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">/=</span> <span class="n">psum</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># select individuals</span>
    <span class="n">chosen</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">k</span><span class="p">):</span>
        <span class="c1">#use a binary search </span>
        <span class="n">cutoff</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span>
        <span class="n">lowerbound</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">upperbound</span> <span class="o">=</span> <span class="n">n</span><span class="o">-</span><span class="mi">1</span>

        <span class="k">while</span><span class="p">(</span><span class="n">upperbound</span> <span class="o">&gt;=</span> <span class="n">lowerbound</span><span class="p">):</span>
            <span class="n">indindex</span><span class="o">=</span><span class="n">lowerbound</span> <span class="o">+</span> <span class="p">(</span><span class="n">upperbound</span><span class="o">-</span><span class="n">lowerbound</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span>
            <span class="k">if</span> <span class="n">psum</span><span class="p">[</span><span class="n">indindex</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">cutoff</span><span class="p">:</span>
                <span class="n">upperbound</span> <span class="o">=</span> <span class="n">indindex</span><span class="o">-</span><span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">lowerbound</span> <span class="o">=</span> <span class="n">indindex</span><span class="o">+</span><span class="mi">1</span>

        <span class="n">lowerbound</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">n</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">lowerbound</span><span class="p">)</span> 
        <span class="n">lowerbound</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">lowerbound</span><span class="p">)</span>

        <span class="n">chosen</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">individuals</span><span class="p">[</span><span class="n">lowerbound</span><span class="p">])</span>    

    <span class="k">return</span> <span class="n">chosen</span>    
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.selAutoEpsilonLexicase" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">selAutoEpsilonLexicase</span><span class="p">(</span><span class="n">individuals</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Adapted from DEAP selAutomaticEpsilonLexicase
Returns an individual that does the best on the fitness cases when considered one at a
time in random order.
https://push-language.hampshire.edu/uploads/default/original/1X/35c30e47ef6323a0a949402914453f277fb1b5b0.pdf
Implemented lambda_epsilon_y implementation.</p>
<p>Args;
    individuals: A list of individuals to select from.
    k: The number of individuals to select.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of selected individuals.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">2239</span>
<span class="normal">2240</span>
<span class="normal">2241</span>
<span class="normal">2242</span>
<span class="normal">2243</span>
<span class="normal">2244</span>
<span class="normal">2245</span>
<span class="normal">2246</span>
<span class="normal">2247</span>
<span class="normal">2248</span>
<span class="normal">2249</span>
<span class="normal">2250</span>
<span class="normal">2251</span>
<span class="normal">2252</span>
<span class="normal">2253</span>
<span class="normal">2254</span>
<span class="normal">2255</span>
<span class="normal">2256</span>
<span class="normal">2257</span>
<span class="normal">2258</span>
<span class="normal">2259</span>
<span class="normal">2260</span>
<span class="normal">2261</span>
<span class="normal">2262</span>
<span class="normal">2263</span>
<span class="normal">2264</span>
<span class="normal">2265</span>
<span class="normal">2266</span>
<span class="normal">2267</span>
<span class="normal">2268</span>
<span class="normal">2269</span>
<span class="normal">2270</span>
<span class="normal">2271</span>
<span class="normal">2272</span>
<span class="normal">2273</span>
<span class="normal">2274</span>
<span class="normal">2275</span>
<span class="normal">2276</span>
<span class="normal">2277</span>
<span class="normal">2278</span>
<span class="normal">2279</span>
<span class="normal">2280</span>
<span class="normal">2281</span>
<span class="normal">2282</span>
<span class="normal">2283</span>
<span class="normal">2284</span>
<span class="normal">2285</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">selAutoEpsilonLexicase</span><span class="p">(</span><span class="n">individuals</span><span class="p">:</span> <span class="n">leap_sensible_initialisation</span><span class="p">,</span> <span class="n">k</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adapted from DEAP selAutomaticEpsilonLexicase</span>
<span class="sd">    Returns an individual that does the best on the fitness cases when considered one at a</span>
<span class="sd">    time in random order.</span>
<span class="sd">    https://push-language.hampshire.edu/uploads/default/original/1X/35c30e47ef6323a0a949402914453f277fb1b5b0.pdf</span>
<span class="sd">    Implemented lambda_epsilon_y implementation.</span>

<span class="sd">    Args;</span>
<span class="sd">        individuals: A list of individuals to select from.</span>
<span class="sd">        k: The number of individuals to select.</span>

<span class="sd">    Returns: </span>
<span class="sd">        A list of selected individuals.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">selected_individuals</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">nscores</span> <span class="o">=</span>  <span class="nb">len</span><span class="p">(</span><span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptscores</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">k</span><span class="p">):</span>
        <span class="n">candidates</span> <span class="o">=</span> <span class="n">individuals</span>
        <span class="n">cases</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">nscores</span><span class="p">))</span>
        <span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">cases</span><span class="p">)</span>

        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">cases</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">candidates</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">errors_for_this_case</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">ptscores</span><span class="p">[</span><span class="n">cases</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">candidates</span><span class="p">])</span>

            <span class="c1"># check if all candidates are nan for this case</span>
            <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">errors_for_this_case</span><span class="p">))</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">candidates</span><span class="p">):</span>
                <span class="k">break</span>
            <span class="n">nan_mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">errors_for_this_case</span><span class="p">)</span>

            <span class="n">median_val</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">errors_for_this_case</span><span class="p">[</span><span class="o">~</span><span class="n">nan_mask</span><span class="p">])</span>
            <span class="n">median_absolute_deviation</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">([</span><span class="nb">abs</span><span class="p">(</span><span class="n">x</span> <span class="o">-</span> <span class="n">median_val</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">errors_for_this_case</span><span class="p">[</span><span class="o">~</span><span class="n">nan_mask</span><span class="p">]])</span>

            <span class="n">best_val_for_case</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">errors_for_this_case</span><span class="p">[</span><span class="o">~</span><span class="n">nan_mask</span><span class="p">])</span>
            <span class="n">max_val_to_survive</span> <span class="o">=</span> <span class="n">best_val_for_case</span> <span class="o">+</span> <span class="n">median_absolute_deviation</span>
            <span class="n">new_candidates</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">candidates</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">ptscores</span><span class="p">[</span><span class="n">cases</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span> <span class="ow">and</span> <span class="n">x</span><span class="o">.</span><span class="n">ptscores</span><span class="p">[</span><span class="n">cases</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">&lt;=</span> <span class="n">max_val_to_survive</span><span class="p">:</span>
                    <span class="n">new_candidates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
            <span class="n">candidates</span> <span class="o">=</span> <span class="n">new_candidates</span>
            <span class="n">cases</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">selected_individuals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">candidates</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">selected_individuals</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.selBalAccLexicase" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">selBalAccLexicase</span><span class="p">(</span><span class="n">individuals</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Returns an individual that does the best on the fitness cases when
considered one at a time in random order.
http://faculty.hampshire.edu/lspector/pubs/lexicase-IEEE-TEC.pdf
Matches will have 0 for a result while mismatches are 1
nan also causes a candidate to drop out</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>individuals</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="grape.grape.leap_sensible_initialisation" href="#grape.grape.leap_sensible_initialisation">leap_sensible_initialisation</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of individuals to select from.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of individuals to select.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of selected individuals.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">2203</span>
<span class="normal">2204</span>
<span class="normal">2205</span>
<span class="normal">2206</span>
<span class="normal">2207</span>
<span class="normal">2208</span>
<span class="normal">2209</span>
<span class="normal">2210</span>
<span class="normal">2211</span>
<span class="normal">2212</span>
<span class="normal">2213</span>
<span class="normal">2214</span>
<span class="normal">2215</span>
<span class="normal">2216</span>
<span class="normal">2217</span>
<span class="normal">2218</span>
<span class="normal">2219</span>
<span class="normal">2220</span>
<span class="normal">2221</span>
<span class="normal">2222</span>
<span class="normal">2223</span>
<span class="normal">2224</span>
<span class="normal">2225</span>
<span class="normal">2226</span>
<span class="normal">2227</span>
<span class="normal">2228</span>
<span class="normal">2229</span>
<span class="normal">2230</span>
<span class="normal">2231</span>
<span class="normal">2232</span>
<span class="normal">2233</span>
<span class="normal">2234</span>
<span class="normal">2235</span>
<span class="normal">2236</span>
<span class="normal">2237</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">selBalAccLexicase</span><span class="p">(</span><span class="n">individuals</span><span class="p">:</span> <span class="n">leap_sensible_initialisation</span><span class="p">,</span> <span class="n">k</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns an individual that does the best on the fitness cases when</span>
<span class="sd">    considered one at a time in random order.</span>
<span class="sd">    http://faculty.hampshire.edu/lspector/pubs/lexicase-IEEE-TEC.pdf</span>
<span class="sd">    Matches will have 0 for a result while mismatches are 1</span>
<span class="sd">    nan also causes a candidate to drop out</span>

<span class="sd">    Args:</span>
<span class="sd">        individuals: A list of individuals to select from.</span>
<span class="sd">        k: The number of individuals to select.</span>

<span class="sd">    Returns: </span>
<span class="sd">        A list of selected individuals.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">selected_individuals</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="n">nscores</span> <span class="o">=</span>  <span class="nb">len</span><span class="p">(</span><span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ptscores</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">k</span><span class="p">):</span>
        <span class="n">candidates</span> <span class="o">=</span> <span class="n">individuals</span>
        <span class="n">cases</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">nscores</span><span class="p">))</span>
        <span class="n">random</span><span class="o">.</span><span class="n">shuffle</span><span class="p">(</span><span class="n">cases</span><span class="p">)</span>

        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">cases</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">candidates</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">newcandidates</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">candidates</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">ptscores</span><span class="p">[</span><span class="n">cases</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">newcandidates</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="n">candidates</span> <span class="o">=</span> <span class="n">newcandidates</span>
            <span class="n">cases</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">selected_individuals</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">candidates</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">selected_individuals</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.selTournRoulette" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">selTournRoulette</span><span class="p">(</span><span class="n">individuals</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">fit_attr</span><span class="o">=</span><span class="s1">&#39;fitness&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Pick two individuals from the population using the RouletteWheel selection
method.  Then return the better of the two individuals. Repeat to get k individuals
to return. Matches the ATHENA tournament selection method</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>individuals</code></td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of individuals to select from.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of individuals to select.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>fit_attr</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The attribute of individuals to use as selection criterion</p>
              </div>
            </td>
            <td>
                  <code>&#39;fitness&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list containing the k selected individuals</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Return:</p>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">2135</span>
<span class="normal">2136</span>
<span class="normal">2137</span>
<span class="normal">2138</span>
<span class="normal">2139</span>
<span class="normal">2140</span>
<span class="normal">2141</span>
<span class="normal">2142</span>
<span class="normal">2143</span>
<span class="normal">2144</span>
<span class="normal">2145</span>
<span class="normal">2146</span>
<span class="normal">2147</span>
<span class="normal">2148</span>
<span class="normal">2149</span>
<span class="normal">2150</span>
<span class="normal">2151</span>
<span class="normal">2152</span>
<span class="normal">2153</span>
<span class="normal">2154</span>
<span class="normal">2155</span>
<span class="normal">2156</span>
<span class="normal">2157</span>
<span class="normal">2158</span>
<span class="normal">2159</span>
<span class="normal">2160</span>
<span class="normal">2161</span>
<span class="normal">2162</span>
<span class="normal">2163</span>
<span class="normal">2164</span>
<span class="normal">2165</span>
<span class="normal">2166</span>
<span class="normal">2167</span>
<span class="normal">2168</span>
<span class="normal">2169</span>
<span class="normal">2170</span>
<span class="normal">2171</span>
<span class="normal">2172</span>
<span class="normal">2173</span>
<span class="normal">2174</span>
<span class="normal">2175</span>
<span class="normal">2176</span>
<span class="normal">2177</span>
<span class="normal">2178</span>
<span class="normal">2179</span>
<span class="normal">2180</span>
<span class="normal">2181</span>
<span class="normal">2182</span>
<span class="normal">2183</span>
<span class="normal">2184</span>
<span class="normal">2185</span>
<span class="normal">2186</span>
<span class="normal">2187</span>
<span class="normal">2188</span>
<span class="normal">2189</span>
<span class="normal">2190</span>
<span class="normal">2191</span>
<span class="normal">2192</span>
<span class="normal">2193</span>
<span class="normal">2194</span>
<span class="normal">2195</span>
<span class="normal">2196</span>
<span class="normal">2197</span>
<span class="normal">2198</span>
<span class="normal">2199</span>
<span class="normal">2200</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">selTournRoulette</span><span class="p">(</span><span class="n">individuals</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">k</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">fit_attr</span><span class="p">:</span><span class="nb">str</span><span class="o">=</span><span class="s2">&quot;fitness&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Pick two individuals from the population using the RouletteWheel selection</span>
<span class="sd">    method.  Then return the better of the two individuals. Repeat to get k individuals</span>
<span class="sd">    to return. Matches the ATHENA tournament selection method</span>

<span class="sd">    Args:</span>
<span class="sd">        individuals: A list of individuals to select from.</span>
<span class="sd">        k: The number of individuals to select.</span>
<span class="sd">        fit_attr: The attribute of individuals to use as selection criterion</span>

<span class="sd">    Returns: </span>
<span class="sd">        A list containing the k selected individuals</span>

<span class="sd">    Return:</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># sort the population in ascending order</span>
    <span class="n">individuals</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">individuals</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="n">fit_attr</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">psum</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">individuals</span><span class="p">)</span>

    <span class="c1"># set the probabilities for each individual to be selected</span>
    <span class="c1"># if all identical then make all probabilities equal</span>
    <span class="k">if</span> <span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">individuals</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
        <span class="n">psum</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">fitness</span><span class="o">/</span><span class="nb">float</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">individuals</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">maxfit</span> <span class="o">=</span> <span class="n">individuals</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">minfit</span> <span class="o">=</span> <span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="c1"># set first probability</span>
        <span class="n">psum</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="o">-</span><span class="n">individuals</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">maxfit</span> <span class="o">+</span> <span class="n">minfit</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">ind</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">individuals</span><span class="p">[</span><span class="mi">1</span><span class="p">:],</span> <span class="n">start</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">psum</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="o">-</span><span class="n">ind</span><span class="o">.</span><span class="n">fitness</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">minfit</span> <span class="o">+</span> <span class="n">maxfit</span> <span class="o">+</span> <span class="n">psum</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
            <span class="n">psum</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">/=</span> <span class="n">psum</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># select individuals</span>
    <span class="n">chosen</span><span class="o">=</span><span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">k</span><span class="p">):</span>
        <span class="c1">#use a binary search </span>
        <span class="n">aspirants</span><span class="o">=</span><span class="p">[]</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
            <span class="n">cutoff</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span>
            <span class="n">lowerbound</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">upperbound</span> <span class="o">=</span> <span class="n">n</span><span class="o">-</span><span class="mi">1</span>

            <span class="k">while</span><span class="p">(</span><span class="n">upperbound</span> <span class="o">&gt;=</span> <span class="n">lowerbound</span><span class="p">):</span>
                <span class="n">indindex</span><span class="o">=</span><span class="n">lowerbound</span> <span class="o">+</span> <span class="p">(</span><span class="n">upperbound</span><span class="o">-</span><span class="n">lowerbound</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span>
                <span class="k">if</span> <span class="n">psum</span><span class="p">[</span><span class="n">indindex</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">cutoff</span><span class="p">:</span>
                    <span class="n">upperbound</span> <span class="o">=</span> <span class="n">indindex</span><span class="o">-</span><span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">lowerbound</span> <span class="o">=</span> <span class="n">indindex</span><span class="o">+</span><span class="mi">1</span>

            <span class="n">lowerbound</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">n</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">lowerbound</span><span class="p">)</span> 
            <span class="n">lowerbound</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">lowerbound</span><span class="p">)</span>

            <span class="n">aspirants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">individuals</span><span class="p">[</span><span class="n">lowerbound</span><span class="p">])</span>

        <span class="n">selected</span> <span class="o">=</span> <span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">aspirants</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="n">fit_attr</span><span class="p">)))</span>

        <span class="n">chosen</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">selected</span><span class="p">)</span>    

    <span class="k">return</span> <span class="n">chosen</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.selTournamentWithoutInvalids" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">selTournamentWithoutInvalids</span><span class="p">(</span><span class="n">individuals</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">tournsize</span><span class="p">,</span> <span class="n">fit_attr</span><span class="o">=</span><span class="s1">&#39;fitness&#39;</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>A simple tournament selection which avoids invalid individuals.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>individuals</code></td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>population to select from</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>k</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number to select</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>tournsize</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>tournament size</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>fit_attr</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>attribute to use in selecting individuals</p>
              </div>
            </td>
            <td>
                  <code>&#39;fitness&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of selected individuals</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">2056</span>
<span class="normal">2057</span>
<span class="normal">2058</span>
<span class="normal">2059</span>
<span class="normal">2060</span>
<span class="normal">2061</span>
<span class="normal">2062</span>
<span class="normal">2063</span>
<span class="normal">2064</span>
<span class="normal">2065</span>
<span class="normal">2066</span>
<span class="normal">2067</span>
<span class="normal">2068</span>
<span class="normal">2069</span>
<span class="normal">2070</span>
<span class="normal">2071</span>
<span class="normal">2072</span>
<span class="normal">2073</span>
<span class="normal">2074</span>
<span class="normal">2075</span>
<span class="normal">2076</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">selTournamentWithoutInvalids</span><span class="p">(</span><span class="n">individuals</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">k</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">tournsize</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> 
                                 <span class="n">fit_attr</span><span class="p">:</span> <span class="nb">str</span><span class="o">=</span><span class="s2">&quot;fitness&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A simple tournament selection which avoids invalid individuals.</span>

<span class="sd">    Args:</span>
<span class="sd">        individuals: population to select from</span>
<span class="sd">        k: number to select</span>
<span class="sd">        tournsize: tournament size</span>
<span class="sd">        fit_attr: attribute to use in selecting individuals</span>

<span class="sd">    Returns:</span>
<span class="sd">        list of selected individuals</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">chosen</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">valid_individuals</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">individuals</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="o">.</span><span class="n">invalid</span><span class="p">]</span>
    <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">chosen</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">k</span><span class="p">:</span>
        <span class="n">aspirants</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">valid_individuals</span><span class="p">,</span> <span class="n">tournsize</span><span class="p">)</span>
        <span class="n">chosen</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">aspirants</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="n">attrgetter</span><span class="p">(</span><span class="n">fit_attr</span><span class="p">)))</span>
    <span class="k">return</span> <span class="n">chosen</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grape.grape.sensible_initialisation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sensible_initialisation</span><span class="p">(</span><span class="n">ind_class</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">min_init_depth</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">,</span> <span class="n">genome_representation</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Sensible initialization for standard linear genome. Half the generated individuals
are up to the depth of the max_init_depth and half are limited to that depth but may be
smaller</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>ind_class</code></td>
            <td>
                  <code><span title="deap.creator.MetaCreator">MetaCreator</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Deap creator</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>pop_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>number of individuals to create</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>bnf_grammar</code></td>
            <td>
                  <code><span title="grape.grape.grape.Grammar">Grammar</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>BNF grammar </p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>min_init_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>minimum depth for tree in mapping genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>max_init_depth</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum depth for tree in mapping genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_size</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>maximum to be stored in codon</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>codon_consumption</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>type of consumption when mapping (ie. lazy or eager)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>genome_representation</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>either list or numpy</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>population</code></td>            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of individuals</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>grape/grape.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">sensible_initialisation</span><span class="p">(</span><span class="n">ind_class</span><span class="p">:</span> <span class="s1">&#39;deap.creator.MetaCreator&#39;</span><span class="p">,</span> <span class="n">pop_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
                          <span class="n">bnf_grammar</span><span class="p">:</span><span class="s1">&#39;grape.grape.Grammar&#39;</span><span class="p">,</span> 
                          <span class="n">min_init_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> 
                          <span class="n">max_init_depth</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_size</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span>
                          <span class="n">genome_representation</span><span class="p">:</span><span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Sensible initialization for standard linear genome. Half the generated individuals</span>
<span class="sd">    are up to the depth of the max_init_depth and half are limited to that depth but may be</span>
<span class="sd">    smaller</span>

<span class="sd">    Args:</span>
<span class="sd">        ind_class: Deap creator</span>
<span class="sd">        pop_size: number of individuals to create</span>
<span class="sd">        bnf_grammar: BNF grammar </span>
<span class="sd">        min_init_depth: minimum depth for tree in mapping genome</span>
<span class="sd">        max_init_depth: maximum depth for tree in mapping genome</span>
<span class="sd">        codon_size: maximum to be stored in codon</span>
<span class="sd">        codon_consumption: type of consumption when mapping (ie. lazy or eager)</span>
<span class="sd">        genome_representation: either list or numpy</span>


<span class="sd">    Returns:</span>
<span class="sd">        population: list of individuals</span>
<span class="sd">    &quot;&quot;&quot;</span> 
    <span class="c1">#Calculate the number of individuals to be generated with each method</span>
    <span class="n">is_odd</span> <span class="o">=</span> <span class="n">pop_size</span> <span class="o">%</span> <span class="mi">2</span>
    <span class="n">n_grow</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pop_size</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span>

    <span class="n">n_sets_grow</span> <span class="o">=</span> <span class="n">max_init_depth</span> <span class="o">-</span> <span class="n">min_init_depth</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">set_size</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">n_grow</span><span class="o">/</span><span class="n">n_sets_grow</span><span class="p">)</span>
    <span class="n">remaining</span> <span class="o">=</span> <span class="n">n_grow</span> <span class="o">%</span> <span class="n">n_sets_grow</span>

    <span class="n">n_full</span> <span class="o">=</span> <span class="n">n_grow</span> <span class="o">+</span> <span class="n">is_odd</span> <span class="o">+</span> <span class="n">remaining</span> <span class="c1">#if pop_size is odd, generate an extra ind with &quot;full&quot;</span>

    <span class="c1">#TODO check if it is possible to generate inds with max_init_depth</span>

    <span class="n">population</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1">#Generate inds using &quot;Grow&quot;</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_sets_grow</span><span class="p">):</span>
        <span class="n">max_init_depth_</span> <span class="o">=</span> <span class="n">min_init_depth</span> <span class="o">+</span> <span class="n">i</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">set_size</span><span class="p">):</span>
            <span class="n">remainders</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the choices</span>
            <span class="n">possible_choices</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the respective possible choices</span>

            <span class="n">phenotype</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">start_rule</span>
            <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span> <span class="c1">#</span>
            <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="c1">#it keeps the depth of each branch</span>
            <span class="n">idx_branch</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1">#index of the current branch being grown</span>
            <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">idx_NT</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">total_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]]</span>
                <span class="n">actual_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]</span> <span class="k">if</span> <span class="n">PR</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_init_depth_</span><span class="p">]</span>
                <span class="n">Ch</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">actual_options</span><span class="p">)</span>
                <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
                <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
                    <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                    <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>
                <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                        <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>

                <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">idx_branch</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">depths</span> <span class="o">=</span> <span class="n">depths</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
                <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;terminal&#39;</span><span class="p">:</span>
                    <span class="n">idx_branch</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span>

            <span class="c1">#Generate the genome</span>
            <span class="n">genome</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span> <span class="ow">or</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">remainders</span><span class="p">)):</span>
                    <span class="n">codon</span> <span class="o">=</span> <span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">int</span><span class="p">(</span><span class="mf">1e10</span><span class="p">))</span> <span class="o">%</span> <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(((</span><span class="n">codon_size</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span> <span class="o">*</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">k</span><span class="p">])</span> <span class="o">+</span> <span class="n">remainders</span><span class="p">[</span><span class="n">k</span><span class="p">]</span>
                    <span class="n">genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codon</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>

            <span class="c1">#Include a tail with 50% of the genome&#39;s size</span>
            <span class="n">size_tail</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.5</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">)),</span> <span class="mi">1</span><span class="p">)</span> <span class="c1">#Tail must have at least one codon. Otherwise, in the lazy approach, when we have the last PR with just a single option, the mapping procces will not terminate.</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">size_tail</span><span class="p">):</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">codon_size</span><span class="p">))</span>

            <span class="c1">#Initialise the individual and include in the population</span>
            <span class="n">ind</span> <span class="o">=</span> <span class="n">ind_class</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_init_depth_</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>

            <span class="c1">#Check if the individual was mapped correctly</span>
<span class="c1">#                 if remainders != ind.structure or phenotype != ind.phenotype or max(depths) != ind.depth:</span>
<span class="c1">#                     print(f&quot;ind.structure={ind.structure} remainders={remainders}&quot;)</span>
<span class="c1">#                     raise Exception(&#39;error in the mapping&#39;)</span>

            <span class="n">population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>    

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_full</span><span class="p">):</span>
        <span class="n">remainders</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the choices</span>
        <span class="n">possible_choices</span> <span class="o">=</span> <span class="p">[]</span> <span class="c1">#it will register the respective possible choices</span>

        <span class="n">phenotype</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">start_rule</span>
        <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span> <span class="c1">#</span>
        <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="c1">#it keeps the depth of each branch</span>
        <span class="n">idx_branch</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1">#index of the current branch being grown</span>

        <span class="k">while</span> <span class="nb">len</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">idx_NT</span> <span class="o">=</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">non_terminals</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">total_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]]</span>
            <span class="n">actual_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">bnf_grammar</span><span class="o">.</span><span class="n">production_rules</span><span class="p">[</span><span class="n">idx_NT</span><span class="p">]</span> <span class="k">if</span> <span class="n">PR</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">max_init_depth</span><span class="p">]</span>
            <span class="n">recursive_options</span> <span class="o">=</span> <span class="p">[</span><span class="n">PR</span> <span class="k">for</span> <span class="n">PR</span> <span class="ow">in</span> <span class="n">actual_options</span> <span class="k">if</span> <span class="n">PR</span><span class="p">[</span><span class="mi">4</span><span class="p">]]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">recursive_options</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">Ch</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">recursive_options</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">Ch</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">actual_options</span><span class="p">)</span>
            <span class="n">phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">remaining_NTs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span><span class="p">:</span>
                <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>
            <span class="k">elif</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">remainders</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Ch</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                    <span class="n">possible_choices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">total_options</span><span class="p">))</span>

            <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">idx_branch</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">depths</span> <span class="o">=</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">depths</span> <span class="o">=</span> <span class="n">depths</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">idx_branch</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="p">],]</span><span class="o">*</span><span class="n">Ch</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">depths</span><span class="p">[</span><span class="n">idx_branch</span><span class="o">+</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">if</span> <span class="n">Ch</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;terminal&#39;</span><span class="p">:</span>
                <span class="n">idx_branch</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="n">remaining_NTs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;&lt;&#39;</span> <span class="o">+</span> <span class="n">term</span> <span class="o">+</span> <span class="s1">&#39;&gt;&#39;</span> <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\&lt;([\(\)\w,-.]+)\&gt;&quot;</span><span class="p">,</span><span class="n">phenotype</span><span class="p">)]</span>

        <span class="c1">#Generate the genome</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span> <span class="ow">or</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">remainders</span><span class="p">)):</span>
                <span class="n">codon</span> <span class="o">=</span> <span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">int</span><span class="p">(</span><span class="mf">1e10</span><span class="p">))</span> <span class="o">%</span> <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(((</span><span class="n">codon_size</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">j</span><span class="p">]))</span> <span class="o">*</span> <span class="n">possible_choices</span><span class="p">[</span><span class="n">j</span><span class="p">])</span> <span class="o">+</span> <span class="n">remainders</span><span class="p">[</span><span class="n">j</span><span class="p">]</span>
                <span class="n">genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">codon</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unknown mapper&quot;</span><span class="p">)</span>

        <span class="c1">#Include a tail with 50% of the genome&#39;s size</span>
        <span class="k">if</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;eager&#39;</span> <span class="ow">or</span> <span class="n">codon_consumption</span> <span class="o">==</span> <span class="s1">&#39;lazy&#39;</span><span class="p">:</span>
            <span class="n">size_tail</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="mf">0.5</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">)),</span> <span class="mi">1</span><span class="p">)</span> <span class="c1">#Tail must have at least one codon. Otherwise, in the lazy approach, when we have the last PR with just a single option, the mapping procces will not terminate.</span>

        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">size_tail</span><span class="p">):</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">codon_size</span><span class="p">))</span>

        <span class="c1">#Initialise the individual and include in the population</span>
        <span class="n">ind</span> <span class="o">=</span> <span class="n">ind_class</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">bnf_grammar</span><span class="p">,</span> <span class="n">max_init_depth</span><span class="p">,</span> <span class="n">codon_consumption</span><span class="p">)</span>

        <span class="c1">#Check if the individual was mapped correctly</span>
<span class="c1">#             if remainders != ind.structure or phenotype != ind.phenotype or max(depths) != ind.depth:</span>
<span class="c1">#                 raise Exception(&#39;error in the mapping&#39;)</span>

        <span class="n">population</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ind</span><span class="p">)</span>    

    <span class="k">if</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;list&#39;</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">elif</span> <span class="n">genome_representation</span> <span class="o">==</span> <span class="s1">&#39;numpy&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ind</span> <span class="ow">in</span> <span class="n">population</span><span class="p">:</span>
            <span class="n">ind</span><span class="o">.</span><span class="n">genome</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">ind</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">population</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Unkonwn genome representation&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.07f07601.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.56dfad97.min.js"></script>
      
    
  </body>
</html>